{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5454e250-40c3-4dd4-8945-62af69f453f5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import scanpy as sc\n",
    "import anndata\n",
    "import numpy as np\n",
    "import gc\n",
    "import pandas as pd \n",
    "pd.set_option('display.max_columns', None)\n",
    "from biomart import BiomartServer\n",
    "from datetime import date\n",
    "import os\n",
    "gc.isenabled()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6fada276-9cc3-4dec-8e5e-8f3934639b69",
   "metadata": {},
   "outputs": [],
   "source": [
    "server = BiomartServer(\"http://www.ensembl.org/biomart\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9e5ab91e-95a3-4b3f-9986-c79251426ae9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'abrachyrhynchus_gene_ensembl': Pink-footed goose genes (ASM259213v1),\n",
      " 'abrachyrhynchus_genomic_sequence': Pink-footed goose sequences (ASM259213v1),\n",
      " 'acalliptera_gene_ensembl': Eastern happy genes (fAstCal1.2),\n",
      " 'acalliptera_genomic_sequence': Eastern happy sequences (fAstCal1.2),\n",
      " 'acarolinensis_gene_ensembl': Green anole genes (AnoCar2.0v2),\n",
      " 'acarolinensis_genomic_sequence': Green anole sequences (AnoCar2.0v2),\n",
      " 'acchrysaetos_gene_ensembl': Golden eagle genes (bAquChr1.2),\n",
      " 'acchrysaetos_genomic_sequence': Golden eagle sequences (bAquChr1.2),\n",
      " 'acitrinellus_gene_ensembl': Midas cichlid genes (Midas_v5),\n",
      " 'acitrinellus_genomic_sequence': Midas cichlid sequences (Midas_v5),\n",
      " 'amelanoleuca_gene_ensembl': Giant panda genes (ASM200744v2),\n",
      " 'amelanoleuca_genomic_sequence': Giant panda sequences (ASM200744v2),\n",
      " 'amexicanus_gene_ensembl': Mexican tetra genes (Astyanax_mexicanus-2.0),\n",
      " 'amexicanus_genomic_sequence': Mexican tetra sequences (Astyanax_mexicanus-2.0),\n",
      " 'anancymaae_gene_ensembl': Ma's night monkey genes (Anan_2.0),\n",
      " 'anancymaae_genomic_sequence': Ma's night monkey sequences (Anan_2.0),\n",
      " 'aocellaris_gene_ensembl': Clown anemonefish genes (AmpOce1.0),\n",
      " 'aocellaris_genomic_sequence': Clown anemonefish sequences (AmpOce1.0),\n",
      " 'apercula_gene_ensembl': Orange clownfish genes (Nemo_v1),\n",
      " 'apercula_genomic_sequence': Orange clownfish sequences (Nemo_v1),\n",
      " 'aplatyrhynchos_gene_ensembl': Mallard genes (ASM874695v1),\n",
      " 'aplatyrhynchos_genomic_sequence': Mallard sequences (ASM874695v1),\n",
      " 'apolyacanthus_gene_ensembl': Spiny chromis genes (ASM210954v1),\n",
      " 'apolyacanthus_genomic_sequence': Spiny chromis sequences (ASM210954v1),\n",
      " 'applatyrhynchos_gene_ensembl': Duck genes (CAU_duck1.0),\n",
      " 'applatyrhynchos_genomic_sequence': Duck sequences (CAU_duck1.0),\n",
      " 'atestudineus_gene_ensembl': Climbing perch genes (fAnaTes1.2),\n",
      " 'atestudineus_genomic_sequence': Climbing perch sequences (fAnaTes1.2),\n",
      " 'bbbison_gene_ensembl': American bison genes (Bison_UMD1.0),\n",
      " 'bbbison_genomic_sequence': American bison sequences (Bison_UMD1.0),\n",
      " 'bgrunniens_gene_ensembl': Domestic yak genes (LU_Bosgru_v3.0),\n",
      " 'bgrunniens_genomic_sequence': Domestic yak sequences (LU_Bosgru_v3.0),\n",
      " 'bihybrid_gene_ensembl': Hybrid - Bos Indicus genes (UOA_Brahman_1),\n",
      " 'bihybrid_genomic_sequence': Hybrid - Bos Indicus sequences (UOA_Brahman_1),\n",
      " 'bmusculus_gene_ensembl': Blue whale genes (mBalMus1.v2),\n",
      " 'bmusculus_genomic_sequence': Blue whale sequences (mBalMus1.v2),\n",
      " 'bmutus_gene_ensembl': Wild yak genes (BosGru_v2.0),\n",
      " 'bmutus_genomic_sequence': Wild yak sequences (BosGru_v2.0),\n",
      " 'bsplendens_gene_ensembl': Siamese fighting fish genes (fBetSpl5.2),\n",
      " 'bsplendens_genomic_sequence': Siamese fighting fish sequences (fBetSpl5.2),\n",
      " 'btaurus_gene_ensembl': Cow genes (ARS-UCD1.2),\n",
      " 'btaurus_genomic_sequence': Cow sequences (ARS-UCD1.2),\n",
      " 'btaurus_qtl_feature': qtl_feature,\n",
      " 'btaurus_snp': Cow Short Variants (SNPs and indels excluding flagged variants) (ARS-UCD1.2),\n",
      " 'btaurus_structvar': Cow Structural Variants (ARS-UCD1.2),\n",
      " 'cabingdonii_gene_ensembl': Abingdon island giant tortoise genes (ASM359739v1),\n",
      " 'cabingdonii_genomic_sequence': Abingdon island giant tortoise sequences (ASM359739v1),\n",
      " 'catys_gene_ensembl': Sooty mangabey genes (Caty_1.0),\n",
      " 'catys_genomic_sequence': Sooty mangabey sequences (Caty_1.0),\n",
      " 'cauratus_gene_ensembl': Goldfish genes (ASM336829v1),\n",
      " 'cauratus_genomic_sequence': Goldfish sequences (ASM336829v1),\n",
      " 'cccarpio_gene_ensembl': Common carp genes (Cypcar_WagV4.0),\n",
      " 'cccarpio_genomic_sequence': Common carp sequences (Cypcar_WagV4.0),\n",
      " 'cdromedarius_gene_ensembl': Arabian camel genes (CamDro2),\n",
      " 'cdromedarius_genomic_sequence': Arabian camel sequences (CamDro2),\n",
      " 'celegans_gene_ensembl': Caenorhabditis elegans (PRJNA13758) genes (WBcel235),\n",
      " 'celegans_genomic_sequence': Caenorhabditis elegans (PRJNA13758) sequences (WBcel235),\n",
      " 'cgchok1gshd_gene_ensembl': Chinese hamster CHOK1GS genes (CHOK1GS_HDv1),\n",
      " 'cgchok1gshd_genomic_sequence': Chinese hamster CHOK1GS sequences (CHOK1GS_HDv1),\n",
      " 'cgobio_gene_ensembl': Channel bull blenny genes (fCotGob3.1),\n",
      " 'cgobio_genomic_sequence': Channel bull blenny sequences (fCotGob3.1),\n",
      " 'charengus_gene_ensembl': Atlantic herring genes (Ch_v2.0.2),\n",
      " 'charengus_genomic_sequence': Atlantic herring sequences (Ch_v2.0.2),\n",
      " 'chircus_gene_ensembl': Goat genes (ARS1),\n",
      " 'chircus_genomic_sequence': Goat sequences (ARS1),\n",
      " 'chircus_qtl_feature': qtl_feature,\n",
      " 'chircus_snp': Goat Short Variants (SNPs and indels excluding flagged variants) (ARS1),\n",
      " 'choffmanni_gene_ensembl': Sloth genes (choHof1),\n",
      " 'choffmanni_genomic_sequence': Sloth sequences (choHof1),\n",
      " 'chyarkandensis_gene_ensembl': Yarkand deer genes (CEY_v1),\n",
      " 'chyarkandensis_genomic_sequence': Yarkand deer sequences (CEY_v1),\n",
      " 'cimitator_gene_ensembl': Panamanian white-faced capuchin genes (Cebus_imitator-1.0),\n",
      " 'cimitator_genomic_sequence': Panamanian white-faced capuchin sequences (Cebus_imitator-1.0),\n",
      " 'cintestinalis_gene_ensembl': C.intestinalis genes (KH),\n",
      " 'cintestinalis_genomic_sequence': C.intestinalis sequences (KH),\n",
      " 'cjacchus_gene_ensembl': White-tufted-ear marmoset genes (mCalJac1.pat.X),\n",
      " 'cjacchus_genomic_sequence': White-tufted-ear marmoset sequences (mCalJac1.pat.X),\n",
      " 'cjaponica_gene_ensembl': Japanese quail genes (Coturnix_japonica_2.0),\n",
      " 'cjaponica_genomic_sequence': Japanese quail sequences (Coturnix_japonica_2.0),\n",
      " 'clanigera_gene_ensembl': Long-tailed chinchilla genes (ChiLan1.0),\n",
      " 'clanigera_genomic_sequence': Long-tailed chinchilla sequences (ChiLan1.0),\n",
      " 'cldingo_gene_ensembl': Dingo genes (ASM325472v1),\n",
      " 'cldingo_genomic_sequence': Dingo sequences (ASM325472v1),\n",
      " 'clfamiliaris_gene_ensembl': Dog genes (ROS_Cfam_1.0),\n",
      " 'clfamiliaris_genomic_sequence': Dog sequences (ROS_Cfam_1.0),\n",
      " 'clfamiliaris_snp': Dog Short Variants (SNPs and indels excluding flagged variants) (ROS_Cfam_1.0),\n",
      " 'clfamiliaris_structvar': Dog Structural Variants (ROS_Cfam_1.0),\n",
      " 'closure_BFO': ontology,\n",
      " 'closure_BTO': ontology,\n",
      " 'closure_CHEBI': ontology,\n",
      " 'closure_CL': ontology,\n",
      " 'closure_ECO': ontology,\n",
      " 'closure_EFO': ontology,\n",
      " 'closure_EO': ontology,\n",
      " 'closure_FYPO': ontology,\n",
      " 'closure_GO': ontology,\n",
      " 'closure_GRO': ontology,\n",
      " 'closure_MOD': ontology,\n",
      " 'closure_OGMS': ontology,\n",
      " 'closure_PATO': ontology,\n",
      " 'closure_PBO': ontology,\n",
      " 'closure_PBQ': ontology,\n",
      " 'closure_PECO': ontology,\n",
      " 'closure_PHI': ontology,\n",
      " 'closure_PO': ontology,\n",
      " 'closure_SO': ontology,\n",
      " 'closure_SO_mini': ontology_mini,\n",
      " 'closure_SO_motif': ontology_motif,\n",
      " 'closure_SO_regulation': ontology_regulation,\n",
      " 'closure_TO': ontology,\n",
      " 'closure_UO': ontology,\n",
      " 'clumpus_gene_ensembl': Lumpfish genes (fCycLum1.pri),\n",
      " 'clumpus_genomic_sequence': Lumpfish sequences (fCycLum1.pri),\n",
      " 'cmilii_gene_ensembl': Elephant shark genes (Callorhinchus_milii-6.1.3),\n",
      " 'cmilii_genomic_sequence': Elephant shark sequences (Callorhinchus_milii-6.1.3),\n",
      " 'cpbellii_gene_ensembl': Painted turtle genes (Chrysemys_picta_bellii-3.0.3),\n",
      " 'cpbellii_genomic_sequence': Painted turtle sequences (Chrysemys_picta_bellii-3.0.3),\n",
      " 'cporcellus_gene_ensembl': Guinea Pig genes (Cavpor3.0),\n",
      " 'cporcellus_genomic_sequence': Guinea Pig sequences (Cavpor3.0),\n",
      " 'cporosus_gene_ensembl': Australian saltwater crocodile genes (CroPor_comp1),\n",
      " 'cporosus_genomic_sequence': Australian saltwater crocodile sequences (CroPor_comp1),\n",
      " 'csabaeus_gene_ensembl': Vervet-AGM genes (ChlSab1.1),\n",
      " 'csabaeus_genomic_sequence': Vervet-AGM sequences (ChlSab1.1),\n",
      " 'csavignyi_gene_ensembl': C.savignyi genes (CSAV 2.0),\n",
      " 'csavignyi_genomic_sequence': C.savignyi sequences (CSAV 2.0),\n",
      " 'csemilaevis_gene_ensembl': Tongue sole genes (Cse_v1.0),\n",
      " 'csemilaevis_genomic_sequence': Tongue sole sequences (Cse_v1.0),\n",
      " 'csyrichta_gene_ensembl': Tarsier genes (Tarsius_syrichta-2.0.1),\n",
      " 'csyrichta_genomic_sequence': Tarsier sequences (Tarsius_syrichta-2.0.1),\n",
      " 'cvariegatus_gene_ensembl': Sheepshead minnow genes (C_variegatus-1.0),\n",
      " 'cvariegatus_genomic_sequence': Sheepshead minnow sequences (C_variegatus-1.0),\n",
      " 'cwagneri_gene_ensembl': Chacoan peccary genes (CatWag_v2_BIUU_UCD),\n",
      " 'cwagneri_genomic_sequence': Chacoan peccary sequences (CatWag_v2_BIUU_UCD),\n",
      " 'dclupeoides_gene_ensembl': Denticle herring genes (fDenClu1.1),\n",
      " 'dclupeoides_genomic_sequence': Denticle herring sequences (fDenClu1.1),\n",
      " 'dlabrax_gene_ensembl': European seabass genes (dlabrax2021),\n",
      " 'dlabrax_genomic_sequence': European seabass sequences (dlabrax2021),\n",
      " 'dleucas_gene_ensembl': Beluga whale genes (ASM228892v3),\n",
      " 'dleucas_genomic_sequence': Beluga whale sequences (ASM228892v3),\n",
      " 'dmelanogaster_gene_ensembl': Drosophila melanogaster (Fruit fly) genes (BDGP6.46),\n",
      " 'dmelanogaster_genomic_sequence': Drosophila melanogaster (Fruit fly) sequences (BDGP6.46),\n",
      " 'dmelanogaster_karyotype_end': karyotype_end,\n",
      " 'dmelanogaster_karyotype_start': karyotype_start,\n",
      " 'dnovemcinctus_gene_ensembl': Armadillo genes (Dasnov3.0),\n",
      " 'dnovemcinctus_genomic_sequence': Armadillo sequences (Dasnov3.0),\n",
      " 'dordii_gene_ensembl': Kangaroo rat genes (Dord_2.0),\n",
      " 'dordii_genomic_sequence': Kangaroo rat sequences (Dord_2.0),\n",
      " 'drerio_gene_ensembl': Zebrafish genes (GRCz11),\n",
      " 'drerio_genomic_sequence': Zebrafish sequences (GRCz11),\n",
      " 'drerio_snp': Zebrafish Short Variants (SNPs and indels excluding flagged variants) (GRCz11),\n",
      " 'drerio_structvar': Zebrafish Structural Variants (GRCz11),\n",
      " 'easinus_gene_ensembl': Donkey genes (ASM1607732v2),\n",
      " 'easinus_genomic_sequence': Donkey sequences (ASM1607732v2),\n",
      " 'eburgeri_gene_ensembl': Hagfish genes (Eburgeri_3.2),\n",
      " 'eburgeri_genomic_sequence': Hagfish sequences (Eburgeri_3.2),\n",
      " 'ecaballus_gene_ensembl': Horse genes (EquCab3.0),\n",
      " 'ecaballus_genomic_sequence': Horse sequences (EquCab3.0),\n",
      " 'ecaballus_qtl_feature': qtl_feature,\n",
      " 'ecaballus_snp': Horse Short Variants (SNPs and indels excluding flagged variants) (EquCab3.0),\n",
      " 'ecaballus_structvar': Horse Structural Variants (EquCab3.0),\n",
      " 'ecalabaricus_gene_ensembl': Reedfish genes (fErpCal1.1),\n",
      " 'ecalabaricus_genomic_sequence': Reedfish sequences (fErpCal1.1),\n",
      " 'eelectricus_gene_ensembl': Electric eel genes (Ee_SOAP_WITH_SSPACE),\n",
      " 'eelectricus_genomic_sequence': Electric eel sequences (Ee_SOAP_WITH_SSPACE),\n",
      " 'eeuropaeus_gene_ensembl': Hedgehog genes (eriEur1),\n",
      " 'eeuropaeus_genomic_sequence': Hedgehog sequences (eriEur1),\n",
      " 'elucius_gene_ensembl': Northern pike genes (Eluc_v4),\n",
      " 'elucius_genomic_sequence': Northern pike sequences (Eluc_v4),\n",
      " 'etelfairi_gene_ensembl': Lesser hedgehog tenrec genes (TENREC),\n",
      " 'etelfairi_genomic_sequence': Lesser hedgehog tenrec sequences (TENREC),\n",
      " 'falbicollis_gene_ensembl': Collared flycatcher genes (FicAlb1.5),\n",
      " 'falbicollis_genomic_sequence': Collared flycatcher sequences (FicAlb1.5),\n",
      " 'fcatus_gene_ensembl': Cat genes (Felis_catus_9.0),\n",
      " 'fcatus_genomic_sequence': Cat sequences (Felis_catus_9.0),\n",
      " 'fcatus_snp': Cat Short Variants (SNPs and indels excluding flagged variants) (Felis_catus_9.0),\n",
      " 'fheteroclitus_gene_ensembl': Mummichog genes (Fundulus_heteroclitus-3.0.2),\n",
      " 'fheteroclitus_genomic_sequence': Mummichog sequences (Fundulus_heteroclitus-3.0.2),\n",
      " 'gaculeatus_gene_ensembl': Stickleback genes (BROAD S1),\n",
      " 'gaculeatus_genomic_sequence': Stickleback sequences (BROAD S1),\n",
      " 'gaculeatus_marker_end': marker_feature_end,\n",
      " 'gaculeatus_marker_start': marker_feature,\n",
      " 'gevgoodei_gene_ensembl': Goodes thornscrub tortoise genes (rGopEvg1_v1.p),\n",
      " 'gevgoodei_genomic_sequence': Goodes thornscrub tortoise sequences (rGopEvg1_v1.p),\n",
      " 'gfortis_gene_ensembl': Medium ground-finch genes (GeoFor_1.0),\n",
      " 'gfortis_genomic_sequence': Medium ground-finch sequences (GeoFor_1.0),\n",
      " 'ggallus_gene_ensembl': Chicken genes (bGalGal1.mat.broiler.GRCg7b),\n",
      " 'ggallus_genomic_sequence': Chicken sequences (bGalGal1.mat.broiler.GRCg7b),\n",
      " 'ggallus_qtl_feature': qtl_feature,\n",
      " 'ggallus_snp': Chicken Short Variants (SNPs and indels excluding flagged variants) (bGalGal1.mat.broiler.GRCg7b),\n",
      " 'ggorilla_gene_ensembl': Gorilla genes (gorGor4),\n",
      " 'ggorilla_genomic_sequence': Gorilla sequences (gorGor4),\n",
      " 'gmorhua_gene_ensembl': Atlantic cod genes (gadMor3.0),\n",
      " 'gmorhua_genomic_sequence': Atlantic cod sequences (gadMor3.0),\n",
      " 'hburtoni_gene_ensembl': Burton's mouthbrooder genes (AstBur1.0),\n",
      " 'hburtoni_genomic_sequence': Burton's mouthbrooder sequences (AstBur1.0),\n",
      " 'hcomes_gene_ensembl': Tiger tail seahorse genes (H_comes_QL1_v1),\n",
      " 'hcomes_genomic_sequence': Tiger tail seahorse sequences (H_comes_QL1_v1),\n",
      " 'hgfemale_gene_ensembl': Naked mole-rat female genes (Naked_mole-rat_maternal),\n",
      " 'hgfemale_genomic_sequence': Naked mole-rat female sequences (Naked_mole-rat_maternal),\n",
      " 'hhucho_gene_ensembl': Huchen genes (ASM331708v1),\n",
      " 'hhucho_genomic_sequence': Huchen sequences (ASM331708v1),\n",
      " 'hsapiens_encode': encode,\n",
      " 'hsapiens_external_feature': Human Other Regulatory Regions (GRCh38.p14),\n",
      " 'hsapiens_gene_ensembl': Human genes (GRCh38.p14),\n",
      " 'hsapiens_genomic_sequence': Human sequences (GRCh38.p14),\n",
      " 'hsapiens_karyotype_end': karyotype_end,\n",
      " 'hsapiens_karyotype_start': karyotype_start,\n",
      " 'hsapiens_marker_end': marker_feature_end,\n",
      " 'hsapiens_marker_start': marker_feature,\n",
      " 'hsapiens_mirna_target_feature': Human miRNA Target Regions (GRCh38.p14),\n",
      " 'hsapiens_regulatory_feature': Human Regulatory Features (GRCh38.p14),\n",
      " 'hsapiens_snp': Human Short Variants (SNPs and indels excluding flagged variants) (GRCh38.p14),\n",
      " 'hsapiens_snp_som': Human Somatic Short Variants (SNPs and indels excluding flagged variants) (GRCh38.p14),\n",
      " 'hsapiens_structvar': Human Structural Variants (GRCh38.p14),\n",
      " 'hsapiens_structvar_som': Human Somatic Structural Variants (GRCh38.p14),\n",
      " 'ipunctatus_gene_ensembl': Channel catfish genes (IpCoco_1.2),\n",
      " 'ipunctatus_genomic_sequence': Channel catfish sequences (IpCoco_1.2),\n",
      " 'itridecemlineatus_gene_ensembl': Squirrel genes (SpeTri2.0),\n",
      " 'itridecemlineatus_genomic_sequence': Squirrel sequences (SpeTri2.0),\n",
      " 'jjaculus_gene_ensembl': Lesser Egyptian jerboa genes (JacJac1.0),\n",
      " 'jjaculus_genomic_sequence': Lesser Egyptian jerboa sequences (JacJac1.0),\n",
      " 'kmarmoratus_gene_ensembl': Mangrove rivulus genes (ASM164957v1),\n",
      " 'kmarmoratus_genomic_sequence': Mangrove rivulus sequences (ASM164957v1),\n",
      " 'lafricana_gene_ensembl': Elephant genes (Loxafr3.0),\n",
      " 'lafricana_genomic_sequence': Elephant sequences (Loxafr3.0),\n",
      " 'lbergylta_gene_ensembl': Ballan wrasse genes (BallGen_V1),\n",
      " 'lbergylta_genomic_sequence': Ballan wrasse sequences (BallGen_V1),\n",
      " 'lcalcarifer_gene_ensembl': Barramundi perch genes (ASB_HGAPassembly_v1),\n",
      " 'lcalcarifer_genomic_sequence': Barramundi perch sequences (ASB_HGAPassembly_v1),\n",
      " 'lchalumnae_gene_ensembl': Coelacanth genes (LatCha1),\n",
      " 'lchalumnae_genomic_sequence': Coelacanth sequences (LatCha1),\n",
      " 'lcrocea_gene_ensembl': Large yellow croaker genes (L_crocea_2.0),\n",
      " 'lcrocea_genomic_sequence': Large yellow croaker sequences (L_crocea_2.0),\n",
      " 'llaticaudata_gene_ensembl': Blue-ringed sea krait genes (latLat_1.0),\n",
      " 'llaticaudata_genomic_sequence': Blue-ringed sea krait sequences (latLat_1.0),\n",
      " 'lleishanense_gene_ensembl': Leishan spiny toad genes (ASM966780v1),\n",
      " 'lleishanense_genomic_sequence': Leishan spiny toad sequences (ASM966780v1),\n",
      " 'loculatus_gene_ensembl': Spotted gar genes (LepOcu1),\n",
      " 'loculatus_genomic_sequence': Spotted gar sequences (LepOcu1),\n",
      " 'marmatus_gene_ensembl': Zig-zag eel genes (fMasArm1.2),\n",
      " 'marmatus_genomic_sequence': Zig-zag eel sequences (fMasArm1.2),\n",
      " 'mauratus_gene_ensembl': Golden Hamster genes (MesAur1.0),\n",
      " 'mauratus_genomic_sequence': Golden Hamster sequences (MesAur1.0),\n",
      " 'mcaroli_gene_ensembl': Ryukyu mouse genes (CAROLI_EIJ_v1.1),\n",
      " 'mcaroli_genomic_sequence': Ryukyu mouse sequences (CAROLI_EIJ_v1.1),\n",
      " 'mdomestica_gene_ensembl': Opossum genes (ASM229v1),\n",
      " 'mdomestica_genomic_sequence': Opossum sequences (ASM229v1),\n",
      " 'mdomestica_snp': Opossum Short Variants (SNPs and indels excluding flagged variants) (ASM229v1),\n",
      " 'mfascicularis_gene_ensembl': Crab-eating macaque genes (Macaca_fascicularis_6.0),\n",
      " 'mfascicularis_genomic_sequence': Crab-eating macaque sequences (Macaca_fascicularis_6.0),\n",
      " 'mgallopavo_gene_ensembl': Turkey genes (Turkey_5.1),\n",
      " 'mgallopavo_genomic_sequence': Turkey sequences (Turkey_5.1),\n",
      " 'mgallopavo_snp': Turkey Short Variants (SNPs and indels excluding flagged variants) (Turkey_5.1),\n",
      " 'mleucophaeus_gene_ensembl': Drill genes (Mleu.le_1.0),\n",
      " 'mleucophaeus_genomic_sequence': Drill sequences (Mleu.le_1.0),\n",
      " 'mlucifugus_gene_ensembl': Microbat genes (Myoluc2.0),\n",
      " 'mlucifugus_genomic_sequence': Microbat sequences (Myoluc2.0),\n",
      " 'mm129s1svimj_gene_ensembl': Mouse 129S1/SvImJ genes (129S1_SvImJ_v1),\n",
      " 'mm129s1svimj_genomic_sequence': Mouse 129S1/SvImJ sequences (129S1_SvImJ_v1),\n",
      " 'mmaj_gene_ensembl': Mouse A/J genes (A_J_v1),\n",
      " 'mmaj_genomic_sequence': Mouse A/J sequences (A_J_v1),\n",
      " 'mmakrj_gene_ensembl': Mouse AKR/J genes (AKR_J_v1),\n",
      " 'mmakrj_genomic_sequence': Mouse AKR/J sequences (AKR_J_v1),\n",
      " 'mmbalbcj_gene_ensembl': Mouse BALB/cJ genes (BALB_cJ_v1),\n",
      " 'mmbalbcj_genomic_sequence': Mouse BALB/cJ sequences (BALB_cJ_v1),\n",
      " 'mmc3hhej_gene_ensembl': Mouse C3H/HeJ genes (C3H_HeJ_v1),\n",
      " 'mmc3hhej_genomic_sequence': Mouse C3H/HeJ sequences (C3H_HeJ_v1),\n",
      " 'mmc57bl6nj_gene_ensembl': Mouse C57BL/6NJ genes (C57BL_6NJ_v1),\n",
      " 'mmc57bl6nj_genomic_sequence': Mouse C57BL/6NJ sequences (C57BL_6NJ_v1),\n",
      " 'mmcasteij_gene_ensembl': Mouse CAST/EiJ genes (CAST_EiJ_v1),\n",
      " 'mmcasteij_genomic_sequence': Mouse CAST/EiJ sequences (CAST_EiJ_v1),\n",
      " 'mmcbaj_gene_ensembl': Mouse CBA/J genes (CBA_J_v1),\n",
      " 'mmcbaj_genomic_sequence': Mouse CBA/J sequences (CBA_J_v1),\n",
      " 'mmdba2j_gene_ensembl': Mouse DBA/2J genes (DBA_2J_v1),\n",
      " 'mmdba2j_genomic_sequence': Mouse DBA/2J sequences (DBA_2J_v1),\n",
      " 'mmfvbnj_gene_ensembl': Mouse FVB/NJ genes (FVB_NJ_v1),\n",
      " 'mmfvbnj_genomic_sequence': Mouse FVB/NJ sequences (FVB_NJ_v1),\n",
      " 'mmlpj_gene_ensembl': Mouse LP/J genes (LP_J_v1),\n",
      " 'mmlpj_genomic_sequence': Mouse LP/J sequences (LP_J_v1),\n",
      " 'mmmarmota_gene_ensembl': Alpine marmot genes (marMar2.1),\n",
      " 'mmmarmota_genomic_sequence': Alpine marmot sequences (marMar2.1),\n",
      " 'mmnodshiltj_gene_ensembl': Mouse NOD/ShiLtJ genes (NOD_ShiLtJ_v1),\n",
      " 'mmnodshiltj_genomic_sequence': Mouse NOD/ShiLtJ sequences (NOD_ShiLtJ_v1),\n",
      " 'mmnzohlltj_gene_ensembl': Mouse NZO/HlLtJ genes (NZO_HlLtJ_v1),\n",
      " 'mmnzohlltj_genomic_sequence': Mouse NZO/HlLtJ sequences (NZO_HlLtJ_v1),\n",
      " 'mmonoceros_gene_ensembl': Narwhal genes (NGI_Narwhal_1),\n",
      " 'mmonoceros_genomic_sequence': Narwhal sequences (NGI_Narwhal_1),\n",
      " 'mmoschiferus_gene_ensembl': Siberian musk deer genes (MosMos_v2_BIUU_UCD),\n",
      " 'mmoschiferus_genomic_sequence': Siberian musk deer sequences (MosMos_v2_BIUU_UCD),\n",
      " 'mmpwkphj_gene_ensembl': Mouse PWK/PhJ genes (PWK_PhJ_v1),\n",
      " 'mmpwkphj_genomic_sequence': Mouse PWK/PhJ sequences (PWK_PhJ_v1),\n",
      " 'mmulatta_gene_ensembl': Macaque genes (Mmul_10),\n",
      " 'mmulatta_genomic_sequence': Macaque sequences (Mmul_10),\n",
      " 'mmurdjan_gene_ensembl': Pinecone soldierfish genes (fMyrMur1.1),\n",
      " 'mmurdjan_genomic_sequence': Pinecone soldierfish sequences (fMyrMur1.1),\n",
      " 'mmurinus_gene_ensembl': Mouse Lemur genes (Mmur_3.0),\n",
      " 'mmurinus_genomic_sequence': Mouse Lemur sequences (Mmur_3.0),\n",
      " 'mmusculus_external_feature': Mouse Other Regulatory Regions (GRCm39),\n",
      " 'mmusculus_gene_ensembl': Mouse genes (GRCm39),\n",
      " 'mmusculus_genomic_sequence': Mouse sequences (GRCm39),\n",
      " 'mmusculus_karyotype_end': karyotype_end,\n",
      " 'mmusculus_karyotype_start': karyotype_start,\n",
      " 'mmusculus_mirna_target_feature': Mouse miRNA Target Regions (GRCm39),\n",
      " 'mmusculus_regulatory_feature': Mouse Regulatory Features (GRCm39),\n",
      " 'mmusculus_snp': Mouse Short Variants (SNPs and indels excluding flagged variants) (GRCm39),\n",
      " 'mmusculus_structvar': Mouse Structural Variants (GRCm39),\n",
      " 'mmwsbeij_gene_ensembl': Mouse WSB/EiJ genes (WSB_EiJ_v1),\n",
      " 'mmwsbeij_genomic_sequence': Mouse WSB/EiJ sequences (WSB_EiJ_v1),\n",
      " 'mnemestrina_gene_ensembl': Pig-tailed macaque genes (Mnem_1.0),\n",
      " 'mnemestrina_genomic_sequence': Pig-tailed macaque sequences (Mnem_1.0),\n",
      " 'mochrogaster_gene_ensembl': Prairie vole genes (MicOch1.0),\n",
      " 'mochrogaster_genomic_sequence': Prairie vole sequences (MicOch1.0),\n",
      " 'mpahari_gene_ensembl': Shrew mouse genes (PAHARI_EIJ_v1.1),\n",
      " 'mpahari_genomic_sequence': Shrew mouse sequences (PAHARI_EIJ_v1.1),\n",
      " 'mpfuro_gene_ensembl': Ferret genes (MusPutFur1.0),\n",
      " 'mpfuro_genomic_sequence': Ferret sequences (MusPutFur1.0),\n",
      " 'mspicilegus_gene_ensembl': Steppe mouse genes (MUSP714),\n",
      " 'mspicilegus_genomic_sequence': Steppe mouse sequences (MUSP714),\n",
      " 'mspretus_gene_ensembl': Algerian mouse genes (SPRET_EiJ_v1),\n",
      " 'mspretus_genomic_sequence': Algerian mouse sequences (SPRET_EiJ_v1),\n",
      " 'mzebra_gene_ensembl': Zebra mbuna genes (M_zebra_UMD2a),\n",
      " 'mzebra_genomic_sequence': Zebra mbuna sequences (M_zebra_UMD2a),\n",
      " 'nbrichardi_gene_ensembl': Lyretail cichlid genes (NeoBri1.0),\n",
      " 'nbrichardi_genomic_sequence': Lyretail cichlid sequences (NeoBri1.0),\n",
      " 'neugenii_gene_ensembl': Wallaby genes (Meug_1.0),\n",
      " 'neugenii_genomic_sequence': Wallaby sequences (Meug_1.0),\n",
      " 'nfurzeri_gene_ensembl': Turquoise killifish genes (Nfu_20140520),\n",
      " 'nfurzeri_genomic_sequence': Turquoise killifish sequences (Nfu_20140520),\n",
      " 'ngalili_gene_ensembl': Upper Galilee mountains blind mole rat genes (S.galili_v1.0),\n",
      " 'ngalili_genomic_sequence': Upper Galilee mountains blind mole rat sequences (S.galili_v1.0),\n",
      " 'nleucogenys_gene_ensembl': Gibbon genes (Nleu_3.0),\n",
      " 'nleucogenys_genomic_sequence': Gibbon sequences (Nleu_3.0),\n",
      " 'nleucogenys_snp': Gibbon Short Variants (SNPs and indels excluding flagged variants) (Nleu_3.0),\n",
      " 'nnaja_gene_ensembl': Indian cobra genes (Nana_v5),\n",
      " 'nnaja_genomic_sequence': Indian cobra sequences (Nana_v5),\n",
      " 'nscutatus_gene_ensembl': Mainland tiger snake genes (TS10Xv2-PRI),\n",
      " 'nscutatus_genomic_sequence': Mainland tiger snake sequences (TS10Xv2-PRI),\n",
      " 'nvison_gene_ensembl': American mink genes (NNQGG.v01),\n",
      " 'nvison_genomic_sequence': American mink sequences (NNQGG.v01),\n",
      " 'oanatinus_gene_ensembl': Platypus genes (mOrnAna1.p.v1),\n",
      " 'oanatinus_genomic_sequence': Platypus sequences (mOrnAna1.p.v1),\n",
      " 'oanatinus_snp': Platypus Short Variants (SNPs and indels excluding flagged variants) (mOrnAna1.p.v1),\n",
      " 'oarambouillet_gene_ensembl': Sheep genes (Oar_rambouillet_v1.0),\n",
      " 'oarambouillet_genomic_sequence': Sheep sequences (Oar_rambouillet_v1.0),\n",
      " 'oarambouillet_qtl_feature': qtl_feature,\n",
      " 'oarambouillet_snp': Sheep Short Variants (SNPs and indels excluding flagged variants) (Oar_rambouillet_v1.0),\n",
      " 'oarambouillet_structvar': Sheep Structural Variants (Oar_rambouillet_v1.0),\n",
      " 'oaries_gene_ensembl': Sheep (texel) genes (Oar_v3.1),\n",
      " 'oaries_genomic_sequence': Sheep (texel) sequences (Oar_v3.1),\n",
      " 'oaries_marker_end': marker_feature_end,\n",
      " 'oaries_marker_start': marker_feature,\n",
      " 'oaries_qtl_feature': qtl_feature,\n",
      " 'oaries_snp': Sheep (texel) Short Variants (SNPs and indels excluding flagged variants) (Oar_v3.1),\n",
      " 'oaries_structvar': Sheep (texel) Structural Variants (Oar_v3.1),\n",
      " 'ocuniculus_gene_ensembl': Rabbit genes (OryCun2.0),\n",
      " 'ocuniculus_genomic_sequence': Rabbit sequences (OryCun2.0),\n",
      " 'odegus_gene_ensembl': Degu genes (OctDeg1.0),\n",
      " 'odegus_genomic_sequence': Degu sequences (OctDeg1.0),\n",
      " 'ogarnettii_gene_ensembl': Bushbaby genes (OtoGar3),\n",
      " 'ogarnettii_genomic_sequence': Bushbaby sequences (OtoGar3),\n",
      " 'ojavanicus_gene_ensembl': Javanese ricefish genes (OJAV_1.1),\n",
      " 'ojavanicus_genomic_sequence': Javanese ricefish sequences (OJAV_1.1),\n",
      " 'okisutch_gene_ensembl': Coho salmon genes (Okis_V2),\n",
      " 'okisutch_genomic_sequence': Coho salmon sequences (Okis_V2),\n",
      " 'olatipes_gene_ensembl': Japanese medaka HdrR genes (ASM223467v1),\n",
      " 'olatipes_genomic_sequence': Japanese medaka HdrR sequences (ASM223467v1),\n",
      " 'omelastigma_gene_ensembl': Indian medaka genes (Om_v0.7.RACA),\n",
      " 'omelastigma_genomic_sequence': Indian medaka sequences (Om_v0.7.RACA),\n",
      " 'omykiss_gene_ensembl': Rainbow trout genes (USDA_OmykA_1.1),\n",
      " 'omykiss_genomic_sequence': Rainbow trout sequences (USDA_OmykA_1.1),\n",
      " 'oniloticus_gene_ensembl': Nile tilapia genes (O_niloticus_UMD_NMBU),\n",
      " 'oniloticus_genomic_sequence': Nile tilapia sequences (O_niloticus_UMD_NMBU),\n",
      " 'oprinceps_gene_ensembl': Pika genes (OchPri2.0-Ens),\n",
      " 'oprinceps_genomic_sequence': Pika sequences (OchPri2.0-Ens),\n",
      " 'osinensis_gene_ensembl': Chinese medaka genes (ASM858656v1),\n",
      " 'osinensis_genomic_sequence': Chinese medaka sequences (ASM858656v1),\n",
      " 'otshawytscha_gene_ensembl': Chinook salmon genes (Otsh_v1.0),\n",
      " 'otshawytscha_genomic_sequence': Chinook salmon sequences (Otsh_v1.0),\n",
      " 'pabelii_gene_ensembl': Sumatran orangutan genes (Susie_PABv2),\n",
      " 'pabelii_genomic_sequence': Sumatran orangutan sequences (Susie_PABv2),\n",
      " 'panubis_gene_ensembl': Olive baboon genes (Panubis1.0),\n",
      " 'panubis_genomic_sequence': Olive baboon sequences (Panubis1.0),\n",
      " 'pcapensis_gene_ensembl': Hyrax genes (proCap1),\n",
      " 'pcapensis_genomic_sequence': Hyrax sequences (proCap1),\n",
      " 'pcatodon_gene_ensembl': Sperm whale genes (ASM283717v2),\n",
      " 'pcatodon_genomic_sequence': Sperm whale sequences (ASM283717v2),\n",
      " 'pcinereus_gene_ensembl': Koala genes (phaCin_unsw_v4.1),\n",
      " 'pcinereus_genomic_sequence': Koala sequences (phaCin_unsw_v4.1),\n",
      " 'pcoquereli_gene_ensembl': Coquerel's sifaka genes (Pcoq_1.0),\n",
      " 'pcoquereli_genomic_sequence': Coquerel's sifaka sequences (Pcoq_1.0),\n",
      " 'pformosa_gene_ensembl': Amazon molly genes (Poecilia_formosa-5.1.2),\n",
      " 'pformosa_genomic_sequence': Amazon molly sequences (Poecilia_formosa-5.1.2),\n",
      " 'pkingsleyae_gene_ensembl': Paramormyrops kingsleyae genes (PKINGS_0.1),\n",
      " 'pkingsleyae_genomic_sequence': Paramormyrops kingsleyae sequences (PKINGS_0.1),\n",
      " 'platipinna_gene_ensembl': Sailfin molly genes (P_latipinna-1.0),\n",
      " 'platipinna_genomic_sequence': Sailfin molly sequences (P_latipinna-1.0),\n",
      " 'pleo_gene_ensembl': Lion genes (PanLeo1.0),\n",
      " 'pleo_genomic_sequence': Lion sequences (PanLeo1.0),\n",
      " 'pmajor_gene_ensembl': Great Tit genes (Parus_major1.1),\n",
      " 'pmajor_genomic_sequence': Great Tit sequences (Parus_major1.1),\n",
      " 'pmarinus_gene_ensembl': Lamprey genes (Pmarinus_7.0),\n",
      " 'pmarinus_genomic_sequence': Lamprey sequences (Pmarinus_7.0),\n",
      " 'pmbairdii_gene_ensembl': Northern American deer mouse genes (HU_Pman_2.1),\n",
      " 'pmbairdii_genomic_sequence': Northern American deer mouse sequences (HU_Pman_2.1),\n",
      " 'pmuralis_gene_ensembl': Common wall lizard genes (PodMur_1.0),\n",
      " 'pmuralis_genomic_sequence': Common wall lizard sequences (PodMur_1.0),\n",
      " 'pnattereri_gene_ensembl': Red-bellied piranha genes (Pygocentrus_nattereri-1.0.2),\n",
      " 'pnattereri_genomic_sequence': Red-bellied piranha sequences (Pygocentrus_nattereri-1.0.2),\n",
      " 'pnyererei_gene_ensembl': Makobe Island cichlid genes (PunNye1.0),\n",
      " 'pnyererei_genomic_sequence': Makobe Island cichlid sequences (PunNye1.0),\n",
      " 'ppaniscus_gene_ensembl': Bonobo genes (panpan1.1),\n",
      " 'ppaniscus_genomic_sequence': Bonobo sequences (panpan1.1),\n",
      " 'ppardus_gene_ensembl': Leopard genes (PanPar1.0),\n",
      " 'ppardus_genomic_sequence': Leopard sequences (PanPar1.0),\n",
      " 'preticulata_gene_ensembl': Guppy genes (Guppy_female_1.0_MT),\n",
      " 'preticulata_genomic_sequence': Guppy sequences (Guppy_female_1.0_MT),\n",
      " 'psimus_gene_ensembl': Greater bamboo lemur genes (Prosim_1.0),\n",
      " 'psimus_genomic_sequence': Greater bamboo lemur sequences (Prosim_1.0),\n",
      " 'psinensis_gene_ensembl': Chinese softshell turtle genes (PelSin_1.0),\n",
      " 'psinensis_genomic_sequence': Chinese softshell turtle sequences (PelSin_1.0),\n",
      " 'psinus_gene_ensembl': Vaquita genes (mPhoSin1.pri),\n",
      " 'psinus_genomic_sequence': Vaquita sequences (mPhoSin1.pri),\n",
      " 'ptaltaica_gene_ensembl': Tiger genes (PanTig1.0),\n",
      " 'ptaltaica_genomic_sequence': Tiger sequences (PanTig1.0),\n",
      " 'ptextilis_gene_ensembl': Eastern brown snake genes (EBS10Xv2-PRI),\n",
      " 'ptextilis_genomic_sequence': Eastern brown snake sequences (EBS10Xv2-PRI),\n",
      " 'ptroglodytes_gene_ensembl': Chimpanzee genes (Pan_tro_3.0),\n",
      " 'ptroglodytes_genomic_sequence': Chimpanzee sequences (Pan_tro_3.0),\n",
      " 'ptroglodytes_snp': Chimpanzee Short Variants (SNPs and indels excluding flagged variants) (Pan_tro_3.0),\n",
      " 'pvampyrus_gene_ensembl': Megabat genes (pteVam1),\n",
      " 'pvampyrus_genomic_sequence': Megabat sequences (pteVam1),\n",
      " 'rbieti_gene_ensembl': Black snub-nosed monkey genes (ASM169854v1),\n",
      " 'rbieti_genomic_sequence': Black snub-nosed monkey sequences (ASM169854v1),\n",
      " 'rferrumequinum_gene_ensembl': Greater horseshoe bat genes (mRhiFer1_v1.p),\n",
      " 'rferrumequinum_genomic_sequence': Greater horseshoe bat sequences (mRhiFer1_v1.p),\n",
      " 'rnorvegicus_gene_ensembl': Rat genes (mRatBN7.2),\n",
      " 'rnorvegicus_genomic_sequence': Rat sequences (mRatBN7.2),\n",
      " 'rnorvegicus_karyotype_end': karyotype_end,\n",
      " 'rnorvegicus_karyotype_start': karyotype_start,\n",
      " 'rnorvegicus_qtl_feature': qtl_feature,\n",
      " 'rnorvegicus_snp': Rat Short Variants (SNPs and indels excluding flagged variants) (mRatBN7.2),\n",
      " 'rroxellana_gene_ensembl': Golden snub-nosed monkey genes (Rrox_v1),\n",
      " 'rroxellana_genomic_sequence': Golden snub-nosed monkey sequences (Rrox_v1),\n",
      " 'saraneus_gene_ensembl': Shrew genes (sorAra1),\n",
      " 'saraneus_genomic_sequence': Shrew sequences (sorAra1),\n",
      " 'saurata_gene_ensembl': Gilthead seabream genes (fSpaAur1.1),\n",
      " 'saurata_genomic_sequence': Gilthead seabream sequences (fSpaAur1.1),\n",
      " 'sbboliviensis_gene_ensembl': Bolivian squirrel monkey genes (SaiBol1.0),\n",
      " 'sbboliviensis_genomic_sequence': Bolivian squirrel monkey sequences (SaiBol1.0),\n",
      " 'scanaria_gene_ensembl': Common canary genes (SCA1),\n",
      " 'scanaria_genomic_sequence': Common canary sequences (SCA1),\n",
      " 'scaustralis_gene_ensembl': African ostrich genes (ASM69896v1),\n",
      " 'scaustralis_genomic_sequence': African ostrich sequences (ASM69896v1),\n",
      " 'scerevisiae_gene_ensembl': Saccharomyces cerevisiae genes (R64-1-1),\n",
      " 'scerevisiae_genomic_sequence': Saccharomyces cerevisiae sequences (R64-1-1),\n",
      " 'scerevisiae_snp': Saccharomyces cerevisiae Short Variants (SNPs and indels excluding flagged variants) (R64-1-1),\n",
      " 'sdumerili_gene_ensembl': Greater amberjack genes (Sdu_1.0),\n",
      " 'sdumerili_genomic_sequence': Greater amberjack sequences (Sdu_1.0),\n",
      " 'sformosus_gene_ensembl': Asian bonytongue genes (fSclFor1.1),\n",
      " 'sformosus_genomic_sequence': Asian bonytongue sequences (fSclFor1.1),\n",
      " 'sgrahami_gene_ensembl': Golden-line barbel genes (SAMN03320097.WGS_v1.1),\n",
      " 'sgrahami_genomic_sequence': Golden-line barbel sequences (SAMN03320097.WGS_v1.1),\n",
      " 'shabroptila_gene_ensembl': Kakapo genes (bStrHab1_v1.p),\n",
      " 'shabroptila_genomic_sequence': Kakapo sequences (bStrHab1_v1.p),\n",
      " 'sharrisii_gene_ensembl': Tasmanian devil genes (mSarHar1.11),\n",
      " 'sharrisii_genomic_sequence': Tasmanian devil sequences (mSarHar1.11),\n",
      " 'sldorsalis_gene_ensembl': Yellowtail amberjack genes (Sedor1),\n",
      " 'sldorsalis_genomic_sequence': Yellowtail amberjack sequences (Sedor1),\n",
      " 'slucioperca_gene_ensembl': Pike-perch genes (SLUC_FBN_1),\n",
      " 'slucioperca_genomic_sequence': Pike-perch sequences (SLUC_FBN_1),\n",
      " 'smaximus_gene_ensembl': Turbot genes (ASM1334776v1),\n",
      " 'smaximus_genomic_sequence': Turbot sequences (ASM1334776v1),\n",
      " 'smerianae_gene_ensembl': Argentine black and white tegu genes (HLtupMer3),\n",
      " 'smerianae_genomic_sequence': Argentine black and white tegu sequences (HLtupMer3),\n",
      " 'spartitus_gene_ensembl': Bicolor damselfish genes (Stegastes_partitus-1.0.2),\n",
      " 'spartitus_genomic_sequence': Bicolor damselfish sequences (Stegastes_partitus-1.0.2),\n",
      " 'spunctatus_gene_ensembl': Tuatara genes (ASM311381v1),\n",
      " 'spunctatus_genomic_sequence': Tuatara sequences (ASM311381v1),\n",
      " 'ssalar_gene_ensembl': Atlantic salmon genes (Ssal_v3.1),\n",
      " 'ssalar_genomic_sequence': Atlantic salmon sequences (Ssal_v3.1),\n",
      " 'ssbamei_gene_ensembl': Pig - Bamei genes (Bamei_pig_v1),\n",
      " 'ssbamei_genomic_sequence': Pig - Bamei sequences (Bamei_pig_v1),\n",
      " 'ssberkshire_gene_ensembl': Pig - Berkshire genes (Berkshire_pig_v1),\n",
      " 'ssberkshire_genomic_sequence': Pig - Berkshire sequences (Berkshire_pig_v1),\n",
      " 'sscrofa_gene_ensembl': Pig genes (Sscrofa11.1),\n",
      " 'sscrofa_genomic_sequence': Pig sequences (Sscrofa11.1),\n",
      " 'sscrofa_qtl_feature': qtl_feature,\n",
      " 'sscrofa_snp': Pig Short Variants (SNPs and indels excluding flagged variants) (Sscrofa11.1),\n",
      " 'sscrofa_structvar': Pig Structural Variants (Sscrofa11.1),\n",
      " 'sshampshire_gene_ensembl': Pig - Hampshire genes (Hampshire_pig_v1),\n",
      " 'sshampshire_genomic_sequence': Pig - Hampshire sequences (Hampshire_pig_v1),\n",
      " 'ssjinhua_gene_ensembl': Pig - Jinhua genes (Jinhua_pig_v1),\n",
      " 'ssjinhua_genomic_sequence': Pig - Jinhua sequences (Jinhua_pig_v1),\n",
      " 'sslandrace_gene_ensembl': Pig - Landrace genes (Landrace_pig_v1),\n",
      " 'sslandrace_genomic_sequence': Pig - Landrace sequences (Landrace_pig_v1),\n",
      " 'sslargewhite_gene_ensembl': Pig - Largewhite genes (Large_White_v1),\n",
      " 'sslargewhite_genomic_sequence': Pig - Largewhite sequences (Large_White_v1),\n",
      " 'ssmeishan_gene_ensembl': Pig - Meishan genes (Meishan_pig_v1),\n",
      " 'ssmeishan_genomic_sequence': Pig - Meishan sequences (Meishan_pig_v1),\n",
      " 'sspietrain_gene_ensembl': Pig - Pietrain genes (Pietrain_pig_v1),\n",
      " 'sspietrain_genomic_sequence': Pig - Pietrain sequences (Pietrain_pig_v1),\n",
      " 'ssrongchang_gene_ensembl': Pig - Rongchang genes (Rongchang_pig_v1),\n",
      " 'ssrongchang_genomic_sequence': Pig - Rongchang sequences (Rongchang_pig_v1),\n",
      " 'sstibetan_gene_ensembl': Pig - Tibetan genes (Tibetan_Pig_v2),\n",
      " 'sstibetan_genomic_sequence': Pig - Tibetan sequences (Tibetan_Pig_v2),\n",
      " 'ssusmarc_gene_ensembl': Pig USMARC genes (USMARCv1.0),\n",
      " 'ssusmarc_genomic_sequence': Pig USMARC sequences (USMARCv1.0),\n",
      " 'sswuzhishan_gene_ensembl': Pig - Wuzhishan genes (minipig_v1.0),\n",
      " 'sswuzhishan_genomic_sequence': Pig - Wuzhishan sequences (minipig_v1.0),\n",
      " 'strutta_gene_ensembl': Brown trout genes (fSalTru1.1),\n",
      " 'strutta_genomic_sequence': Brown trout sequences (fSalTru1.1),\n",
      " 'svulgaris_gene_ensembl': Eurasian red squirrel genes (mSciVul1.1),\n",
      " 'svulgaris_genomic_sequence': Eurasian red squirrel sequences (mSciVul1.1),\n",
      " 'tbelangeri_gene_ensembl': Tree Shrew genes (tupBel1),\n",
      " 'tbelangeri_genomic_sequence': Tree Shrew sequences (tupBel1),\n",
      " 'tctriunguis_gene_ensembl': Three-toed box turtle genes (T_m_triunguis-2.0),\n",
      " 'tctriunguis_genomic_sequence': Three-toed box turtle sequences (T_m_triunguis-2.0),\n",
      " 'tguttata_gene_ensembl': Zebra finch genes (bTaeGut1_v1.p),\n",
      " 'tguttata_genomic_sequence': Zebra finch sequences (bTaeGut1_v1.p),\n",
      " 'tnigroviridis_gene_ensembl': Tetraodon genes (TETRAODON 8.0),\n",
      " 'tnigroviridis_genomic_sequence': Tetraodon sequences (TETRAODON 8.0),\n",
      " 'tnigroviridis_snp': Tetraodon Short Variants (SNPs and indels excluding flagged variants) (TETRAODON 8.0),\n",
      " 'trubripes_gene_ensembl': Fugu genes (fTakRub1.2),\n",
      " 'trubripes_genomic_sequence': Fugu sequences (fTakRub1.2),\n",
      " 'ttruncatus_gene_ensembl': Dolphin genes (turTru1),\n",
      " 'ttruncatus_genomic_sequence': Dolphin sequences (turTru1),\n",
      " 'uamericanus_gene_ensembl': American black bear genes (ASM334442v1),\n",
      " 'uamericanus_genomic_sequence': American black bear sequences (ASM334442v1),\n",
      " 'umaritimus_gene_ensembl': Polar bear genes (UrsMar_1.0),\n",
      " 'umaritimus_genomic_sequence': Polar bear sequences (UrsMar_1.0),\n",
      " 'uparryii_gene_ensembl': Arctic ground squirrel genes (ASM342692v1),\n",
      " 'uparryii_genomic_sequence': Arctic ground squirrel sequences (ASM342692v1),\n",
      " 'vpacos_gene_ensembl': Alpaca genes (vicPac1),\n",
      " 'vpacos_genomic_sequence': Alpaca sequences (vicPac1),\n",
      " 'vursinus_gene_ensembl': Common wombat genes (bare-nosed_wombat_genome_assembly),\n",
      " 'vursinus_genomic_sequence': Common wombat sequences (bare-nosed_wombat_genome_assembly),\n",
      " 'vvulpes_gene_ensembl': Red fox genes (VulVul2.2),\n",
      " 'vvulpes_genomic_sequence': Red fox sequences (VulVul2.2),\n",
      " 'xmaculatus_gene_ensembl': Platyfish genes (X_maculatus-5.0-male),\n",
      " 'xmaculatus_genomic_sequence': Platyfish sequences (X_maculatus-5.0-male),\n",
      " 'xtropicalis_gene_ensembl': Tropical clawed frog genes (UCB_Xtro_10.0),\n",
      " 'xtropicalis_genomic_sequence': Tropical clawed frog sequences (UCB_Xtro_10.0)}\n"
     ]
    }
   ],
   "source": [
    "server.show_datasets()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53108832-dfc1-4e7e-aa63-a36e8ab847a4",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Human query"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "fd92ed3e-c99f-4bf5-b60d-bee3501cac33",
   "metadata": {},
   "outputs": [],
   "source": [
    "ensembl = server.datasets['hsapiens_gene_ensembl']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93636818-5325-45f9-8811-4898fe6d9b3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "ensembl.show_attributes()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "b0dbd229-4017-430b-8464-f3d9c207202d",
   "metadata": {},
   "outputs": [],
   "source": [
    "columns = ['ensembl_gene_id','external_gene_name', 'external_gene_source','entrezgene_accession','entrezgene_id']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "8e347f21-f67d-4e1f-8dac-79a69dd682c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "'definition_1006',\n",
    "'description',\n",
    "'embl',\n",
    "'entrezgene_accession',\n",
    "'entrezgene_id',\n",
    "'entrezgene_trans_name',\n",
    "'external_gene_name',\n",
    "'external_transcript_name',\n",
    "'go_id',\n",
    "'name_1006',\n",
    "'protein_id',\n",
    "'uniprot_gn_id',\n",
    "'uniprot_gn_symbol',\n",
    "'variation_name',\n",
    "'wikigene_id',\n",
    "'wikigene_name']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "8f31f9ca-ffe7-4b90-8fd4-aff517f7746c",
   "metadata": {},
   "outputs": [],
   "source": [
    "response = ensembl.search({'attributes' : columns,\n",
    "                           'mart_instance' : 'ensembl'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "70b0301d-8958-46d8-bfdc-471c52ef3dd1",
   "metadata": {},
   "outputs": [],
   "source": [
    "mapping_df = pd.read_csv(response.url, sep='\\t', header=None, names=columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "dea489db-f8c0-48df-8bbc-c252f7c2a8a1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>external_gene_name</th>\n",
       "      <th>external_gene_source</th>\n",
       "      <th>entrezgene_accession</th>\n",
       "      <th>entrezgene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000210049</td>\n",
       "      <td>MT-TF</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000211459</td>\n",
       "      <td>MT-RNR1</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000210077</td>\n",
       "      <td>MT-TV</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000210082</td>\n",
       "      <td>MT-RNR2</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000209082</td>\n",
       "      <td>MT-TL1</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75714</th>\n",
       "      <td>ENSG00000236500</td>\n",
       "      <td>CD24P1</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75715</th>\n",
       "      <td>ENSG00000197312</td>\n",
       "      <td>DDI2</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>DDI2</td>\n",
       "      <td>84301.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75716</th>\n",
       "      <td>ENSG00000215695</td>\n",
       "      <td>RSC1A1</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>RSC1A1</td>\n",
       "      <td>6248.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75717</th>\n",
       "      <td>ENSG00000271742</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75718</th>\n",
       "      <td>ENSG00000116786</td>\n",
       "      <td>PLEKHM2</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>PLEKHM2</td>\n",
       "      <td>23207.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>75719 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       ensembl_gene_id external_gene_name external_gene_source   \n",
       "0      ENSG00000210049              MT-TF          HGNC Symbol  \\\n",
       "1      ENSG00000211459            MT-RNR1          HGNC Symbol   \n",
       "2      ENSG00000210077              MT-TV          HGNC Symbol   \n",
       "3      ENSG00000210082            MT-RNR2          HGNC Symbol   \n",
       "4      ENSG00000209082             MT-TL1          HGNC Symbol   \n",
       "...                ...                ...                  ...   \n",
       "75714  ENSG00000236500             CD24P1          HGNC Symbol   \n",
       "75715  ENSG00000197312               DDI2          HGNC Symbol   \n",
       "75716  ENSG00000215695             RSC1A1          HGNC Symbol   \n",
       "75717  ENSG00000271742                NaN                  NaN   \n",
       "75718  ENSG00000116786            PLEKHM2          HGNC Symbol   \n",
       "\n",
       "      entrezgene_accession  entrezgene_id  \n",
       "0                      NaN            NaN  \n",
       "1                      NaN            NaN  \n",
       "2                      NaN            NaN  \n",
       "3                      NaN            NaN  \n",
       "4                      NaN            NaN  \n",
       "...                    ...            ...  \n",
       "75714                  NaN            NaN  \n",
       "75715                 DDI2        84301.0  \n",
       "75716               RSC1A1         6248.0  \n",
       "75717                  NaN            NaN  \n",
       "75718              PLEKHM2        23207.0  \n",
       "\n",
       "[75719 rows x 5 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "753e09c6-01c1-4019-add9-d3951b22450f",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Load human data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "66ad764b-2208-4c92-a6d9-dcf26e5eeb1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "human = sc.read_h5ad(os.path.join(os.getcwd(),'Vignettes/frog_zebrafish_embryogenesis/data/human.h5ad'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "id": "78628f2e-0c84-4d0f-b794-05151de116a5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>entrez_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>mouse_homologenes</th>\n",
       "      <th>n_cells</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3.8-1.3</td>\n",
       "      <td>6</td>\n",
       "      <td>353008</td>\n",
       "      <td>HLA complex group 26 (non-protein coding) pseu...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1470</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>A2M-AS1</td>\n",
       "      <td>12</td>\n",
       "      <td>144571</td>\n",
       "      <td>A2M antisense RNA 1 (head to head)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>A2MP1</td>\n",
       "      <td>12</td>\n",
       "      <td>3</td>\n",
       "      <td>alpha-2-macroglobulin pseudogene 1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>1</td>\n",
       "      <td>127550</td>\n",
       "      <td>alpha 1,3-galactosyltransferase 2</td>\n",
       "      <td>A3galt2</td>\n",
       "      <td>158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>A4GNT</td>\n",
       "      <td>3</td>\n",
       "      <td>51146</td>\n",
       "      <td>alpha-1,4-N-acetylglucosaminyltransferase</td>\n",
       "      <td>A4gnt</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50270</th>\n",
       "      <td>ZXDA</td>\n",
       "      <td>X</td>\n",
       "      <td>7789</td>\n",
       "      <td>zinc finger, X-linked, duplicated A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11839</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50271</th>\n",
       "      <td>ZXDB</td>\n",
       "      <td>X</td>\n",
       "      <td>158586</td>\n",
       "      <td>zinc finger, X-linked, duplicated B</td>\n",
       "      <td>Zxdb</td>\n",
       "      <td>64061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50275</th>\n",
       "      <td>ZYG11B</td>\n",
       "      <td>1</td>\n",
       "      <td>79699</td>\n",
       "      <td>zyg-11 family member B, cell cycle regulator</td>\n",
       "      <td>Zyg11b</td>\n",
       "      <td>35581</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50276</th>\n",
       "      <td>ZYX</td>\n",
       "      <td>7</td>\n",
       "      <td>7791</td>\n",
       "      <td>zyxin</td>\n",
       "      <td>Zyx</td>\n",
       "      <td>1950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50278</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>1</td>\n",
       "      <td>26009</td>\n",
       "      <td>zinc finger, ZZ-type containing 3</td>\n",
       "      <td>Zzz3</td>\n",
       "      <td>65196</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>31551 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          gene chromosome  entrez_id   \n",
       "1      3.8-1.3          6     353008  \\\n",
       "9      A2M-AS1         12     144571   \n",
       "11       A2MP1         12          3   \n",
       "12     A3GALT2          1     127550   \n",
       "14       A4GNT          3      51146   \n",
       "...        ...        ...        ...   \n",
       "50270     ZXDA          X       7789   \n",
       "50271     ZXDB          X     158586   \n",
       "50275   ZYG11B          1      79699   \n",
       "50276      ZYX          7       7791   \n",
       "50278     ZZZ3          1      26009   \n",
       "\n",
       "                                               gene_name mouse_homologenes   \n",
       "1      HLA complex group 26 (non-protein coding) pseu...               NaN  \\\n",
       "9                     A2M antisense RNA 1 (head to head)               NaN   \n",
       "11                    alpha-2-macroglobulin pseudogene 1               NaN   \n",
       "12                     alpha 1,3-galactosyltransferase 2           A3galt2   \n",
       "14             alpha-1,4-N-acetylglucosaminyltransferase             A4gnt   \n",
       "...                                                  ...               ...   \n",
       "50270                zinc finger, X-linked, duplicated A               NaN   \n",
       "50271                zinc finger, X-linked, duplicated B              Zxdb   \n",
       "50275       zyg-11 family member B, cell cycle regulator            Zyg11b   \n",
       "50276                                              zyxin               Zyx   \n",
       "50278                  zinc finger, ZZ-type containing 3              Zzz3   \n",
       "\n",
       "       n_cells  \n",
       "1         1470  \n",
       "9          197  \n",
       "11        1727  \n",
       "12         158  \n",
       "14          17  \n",
       "...        ...  \n",
       "50270    11839  \n",
       "50271    64061  \n",
       "50275    35581  \n",
       "50276     1950  \n",
       "50278    65196  \n",
       "\n",
       "[31551 rows x 6 columns]"
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "human.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "890d0aa9-437a-4b62-90de-5f25c6ece366",
   "metadata": {},
   "outputs": [],
   "source": [
    "ncbi_ids = mapping_df[mapping_df.entrezgene_id.isin(human.var['entrez_id'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "7b5fe6c9-89bc-4cb2-b17c-f1f0977ce735",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>external_gene_name</th>\n",
       "      <th>external_gene_source</th>\n",
       "      <th>entrezgene_accession</th>\n",
       "      <th>entrezgene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSG00000198888</td>\n",
       "      <td>MT-ND1</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>ND1</td>\n",
       "      <td>4535.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENSG00000198763</td>\n",
       "      <td>MT-ND2</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>ND2</td>\n",
       "      <td>4536.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENSG00000198804</td>\n",
       "      <td>MT-CO1</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>COX1</td>\n",
       "      <td>4512.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ENSG00000228253</td>\n",
       "      <td>MT-ATP8</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>ATP8</td>\n",
       "      <td>4509.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>ENSG00000198938</td>\n",
       "      <td>MT-CO3</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>COX3</td>\n",
       "      <td>4514.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75704</th>\n",
       "      <td>ENSG00000142615</td>\n",
       "      <td>CELA2A</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>CELA2A</td>\n",
       "      <td>63036.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75705</th>\n",
       "      <td>ENSG00000215704</td>\n",
       "      <td>CELA2B</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>CELA2B</td>\n",
       "      <td>51032.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75706</th>\n",
       "      <td>ENSG00000132906</td>\n",
       "      <td>CASP9</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>CASP9</td>\n",
       "      <td>842.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75716</th>\n",
       "      <td>ENSG00000215695</td>\n",
       "      <td>RSC1A1</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>RSC1A1</td>\n",
       "      <td>6248.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75718</th>\n",
       "      <td>ENSG00000116786</td>\n",
       "      <td>PLEKHM2</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>PLEKHM2</td>\n",
       "      <td>23207.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18261 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       ensembl_gene_id external_gene_name external_gene_source   \n",
       "5      ENSG00000198888             MT-ND1          HGNC Symbol  \\\n",
       "9      ENSG00000198763             MT-ND2          HGNC Symbol   \n",
       "15     ENSG00000198804             MT-CO1          HGNC Symbol   \n",
       "20     ENSG00000228253            MT-ATP8          HGNC Symbol   \n",
       "22     ENSG00000198938             MT-CO3          HGNC Symbol   \n",
       "...                ...                ...                  ...   \n",
       "75704  ENSG00000142615             CELA2A          HGNC Symbol   \n",
       "75705  ENSG00000215704             CELA2B          HGNC Symbol   \n",
       "75706  ENSG00000132906              CASP9          HGNC Symbol   \n",
       "75716  ENSG00000215695             RSC1A1          HGNC Symbol   \n",
       "75718  ENSG00000116786            PLEKHM2          HGNC Symbol   \n",
       "\n",
       "      entrezgene_accession  entrezgene_id  \n",
       "5                      ND1         4535.0  \n",
       "9                      ND2         4536.0  \n",
       "15                    COX1         4512.0  \n",
       "20                    ATP8         4509.0  \n",
       "22                    COX3         4514.0  \n",
       "...                    ...            ...  \n",
       "75704               CELA2A        63036.0  \n",
       "75705               CELA2B        51032.0  \n",
       "75706                CASP9          842.0  \n",
       "75716               RSC1A1         6248.0  \n",
       "75718              PLEKHM2        23207.0  \n",
       "\n",
       "[18261 rows x 5 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ncbi_ids"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d17df93-70fb-48c4-b1ac-29428161bfad",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Remove duplicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "89f5e0f2-61c1-49e5-a927-901d10b565b3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>external_gene_name</th>\n",
       "      <th>external_gene_source</th>\n",
       "      <th>entrezgene_accession</th>\n",
       "      <th>entrezgene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>944</th>\n",
       "      <td>ENSG00000282566</td>\n",
       "      <td>PRAMEF25</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>PRAMEF26</td>\n",
       "      <td>645359.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1039</th>\n",
       "      <td>ENSG00000114374</td>\n",
       "      <td>USP9Y</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>TTTY15</td>\n",
       "      <td>64595.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1197</th>\n",
       "      <td>ENSG00000228240</td>\n",
       "      <td>TTTY17A</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>TTTY17C</td>\n",
       "      <td>474152.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1621</th>\n",
       "      <td>ENSG00000227439</td>\n",
       "      <td>TTTY17B</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>TTTY17C</td>\n",
       "      <td>474152.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2291</th>\n",
       "      <td>ENSG00000291031</td>\n",
       "      <td>BCORP1</td>\n",
       "      <td>NCBI gene (formerly Entrezgene)</td>\n",
       "      <td>LOC105377223</td>\n",
       "      <td>105377223.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75022</th>\n",
       "      <td>ENSG00000143702</td>\n",
       "      <td>CEP170</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>CEP170P1</td>\n",
       "      <td>645455.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75188</th>\n",
       "      <td>ENSG00000290825</td>\n",
       "      <td>DDX11L2</td>\n",
       "      <td>NCBI gene (formerly Entrezgene)</td>\n",
       "      <td>DDX11L16</td>\n",
       "      <td>727856.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75189</th>\n",
       "      <td>ENSG00000290825</td>\n",
       "      <td>DDX11L2</td>\n",
       "      <td>NCBI gene (formerly Entrezgene)</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>100287102.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75190</th>\n",
       "      <td>ENSG00000290825</td>\n",
       "      <td>DDX11L2</td>\n",
       "      <td>NCBI gene (formerly Entrezgene)</td>\n",
       "      <td>DDX11L5</td>\n",
       "      <td>100287596.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75270</th>\n",
       "      <td>ENSG00000158747</td>\n",
       "      <td>NBL1</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>MICOS10-NBL1</td>\n",
       "      <td>100532736.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       ensembl_gene_id external_gene_name             external_gene_source   \n",
       "944    ENSG00000282566           PRAMEF25                      HGNC Symbol  \\\n",
       "1039   ENSG00000114374              USP9Y                      HGNC Symbol   \n",
       "1197   ENSG00000228240            TTTY17A                      HGNC Symbol   \n",
       "1621   ENSG00000227439            TTTY17B                      HGNC Symbol   \n",
       "2291   ENSG00000291031             BCORP1  NCBI gene (formerly Entrezgene)   \n",
       "...                ...                ...                              ...   \n",
       "75022  ENSG00000143702             CEP170                      HGNC Symbol   \n",
       "75188  ENSG00000290825            DDX11L2  NCBI gene (formerly Entrezgene)   \n",
       "75189  ENSG00000290825            DDX11L2  NCBI gene (formerly Entrezgene)   \n",
       "75190  ENSG00000290825            DDX11L2  NCBI gene (formerly Entrezgene)   \n",
       "75270  ENSG00000158747               NBL1                      HGNC Symbol   \n",
       "\n",
       "      entrezgene_accession  entrezgene_id  \n",
       "944               PRAMEF26       645359.0  \n",
       "1039                TTTY15        64595.0  \n",
       "1197               TTTY17C       474152.0  \n",
       "1621               TTTY17C       474152.0  \n",
       "2291          LOC105377223    105377223.0  \n",
       "...                    ...            ...  \n",
       "75022             CEP170P1       645455.0  \n",
       "75188             DDX11L16       727856.0  \n",
       "75189              DDX11L1    100287102.0  \n",
       "75190              DDX11L5    100287596.0  \n",
       "75270         MICOS10-NBL1    100532736.0  \n",
       "\n",
       "[100 rows x 5 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ncbi_ids[ncbi_ids.ensembl_gene_id.duplicated()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "7c1de4c0-c904-45b8-8f70-77b921e9d22c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>external_gene_name</th>\n",
       "      <th>external_gene_source</th>\n",
       "      <th>entrezgene_accession</th>\n",
       "      <th>entrezgene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>949</th>\n",
       "      <td>ENSG00000281987</td>\n",
       "      <td>PRAMEF26</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>PRAMEF26</td>\n",
       "      <td>645359.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1621</th>\n",
       "      <td>ENSG00000227439</td>\n",
       "      <td>TTTY17B</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>TTTY17C</td>\n",
       "      <td>474152.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1671</th>\n",
       "      <td>ENSG00000238074</td>\n",
       "      <td>TSPY9</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>TSPY9</td>\n",
       "      <td>728132.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2305</th>\n",
       "      <td>ENSG00000233803</td>\n",
       "      <td>TSPY4</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>TSPY4</td>\n",
       "      <td>728395.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2318</th>\n",
       "      <td>ENSG00000225560</td>\n",
       "      <td>FAM197Y8</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>FAM197Y8</td>\n",
       "      <td>105379268.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75244</th>\n",
       "      <td>ENSG00000291217</td>\n",
       "      <td>LINC01347</td>\n",
       "      <td>NCBI gene (formerly Entrezgene)</td>\n",
       "      <td>LINC01347</td>\n",
       "      <td>731275.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75285</th>\n",
       "      <td>ENSG00000238142</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOC105376805</td>\n",
       "      <td>105376805.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75376</th>\n",
       "      <td>ENSG00000160767</td>\n",
       "      <td>ENTREP3</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>ENTREP3</td>\n",
       "      <td>10712.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75479</th>\n",
       "      <td>ENSG00000290842</td>\n",
       "      <td>CROCCP3</td>\n",
       "      <td>NCBI gene (formerly Entrezgene)</td>\n",
       "      <td>CROCCP3</td>\n",
       "      <td>114819.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75700</th>\n",
       "      <td>ENSG00000160818</td>\n",
       "      <td>GPATCH4</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>GPATCH4</td>\n",
       "      <td>54865.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2271 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       ensembl_gene_id external_gene_name             external_gene_source   \n",
       "949    ENSG00000281987           PRAMEF26                      HGNC Symbol  \\\n",
       "1621   ENSG00000227439            TTTY17B                      HGNC Symbol   \n",
       "1671   ENSG00000238074              TSPY9                      HGNC Symbol   \n",
       "2305   ENSG00000233803              TSPY4                      HGNC Symbol   \n",
       "2318   ENSG00000225560           FAM197Y8                      HGNC Symbol   \n",
       "...                ...                ...                              ...   \n",
       "75244  ENSG00000291217          LINC01347  NCBI gene (formerly Entrezgene)   \n",
       "75285  ENSG00000238142                NaN                              NaN   \n",
       "75376  ENSG00000160767            ENTREP3                      HGNC Symbol   \n",
       "75479  ENSG00000290842            CROCCP3  NCBI gene (formerly Entrezgene)   \n",
       "75700  ENSG00000160818            GPATCH4                      HGNC Symbol   \n",
       "\n",
       "      entrezgene_accession  entrezgene_id  \n",
       "949               PRAMEF26       645359.0  \n",
       "1621               TTTY17C       474152.0  \n",
       "1671                 TSPY9       728132.0  \n",
       "2305                 TSPY4       728395.0  \n",
       "2318              FAM197Y8    105379268.0  \n",
       "...                    ...            ...  \n",
       "75244            LINC01347       731275.0  \n",
       "75285         LOC105376805    105376805.0  \n",
       "75376              ENTREP3        10712.0  \n",
       "75479              CROCCP3       114819.0  \n",
       "75700              GPATCH4        54865.0  \n",
       "\n",
       "[2271 rows x 5 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ncbi_ids[ncbi_ids.entrezgene_id.duplicated()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "79d71734-d1cb-4707-84b6-d71a9fd0a182",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['HGNC Symbol', nan, 'NCBI gene (formerly Entrezgene)'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ncbi_ids.external_gene_source.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "86601ff7-45f1-4bf9-b145-9123c8675f8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "ncbi_ids = ncbi_ids[ncbi_ids.external_gene_source.isin(['HGNC Symbol'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "7bd3ae5c-5d64-426f-98e5-bd74f1d6fa25",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "90"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ncbi_ids.ensembl_gene_id.duplicated().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "b80afe0d-f260-4803-bed7-aa215794c826",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2170"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ncbi_ids.entrezgene_id.duplicated().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "78986f9b-10b8-4482-8242-7ec22c04a390",
   "metadata": {},
   "outputs": [],
   "source": [
    "duplicates = ncbi_ids[ncbi_ids.ensembl_gene_id.isin(ncbi_ids[ncbi_ids.ensembl_gene_id.duplicated()].ensembl_gene_id)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "91a8f767-8fa4-44d6-9074-8ae86029023b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>external_gene_name</th>\n",
       "      <th>external_gene_source</th>\n",
       "      <th>entrezgene_accession</th>\n",
       "      <th>entrezgene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>943</th>\n",
       "      <td>ENSG00000282566</td>\n",
       "      <td>PRAMEF25</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>PRAMEF25</td>\n",
       "      <td>441873.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>944</th>\n",
       "      <td>ENSG00000282566</td>\n",
       "      <td>PRAMEF25</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>PRAMEF26</td>\n",
       "      <td>645359.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1038</th>\n",
       "      <td>ENSG00000114374</td>\n",
       "      <td>USP9Y</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>USP9Y</td>\n",
       "      <td>8287.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1039</th>\n",
       "      <td>ENSG00000114374</td>\n",
       "      <td>USP9Y</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>TTTY15</td>\n",
       "      <td>64595.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1196</th>\n",
       "      <td>ENSG00000228240</td>\n",
       "      <td>TTTY17A</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>TTTY17A</td>\n",
       "      <td>252949.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74592</th>\n",
       "      <td>ENSG00000229571</td>\n",
       "      <td>PRAMEF25</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>PRAMEF26</td>\n",
       "      <td>645359.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75021</th>\n",
       "      <td>ENSG00000143702</td>\n",
       "      <td>CEP170</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>CEP170</td>\n",
       "      <td>9859.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75022</th>\n",
       "      <td>ENSG00000143702</td>\n",
       "      <td>CEP170</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>CEP170P1</td>\n",
       "      <td>645455.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75269</th>\n",
       "      <td>ENSG00000158747</td>\n",
       "      <td>NBL1</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>NBL1</td>\n",
       "      <td>4681.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75270</th>\n",
       "      <td>ENSG00000158747</td>\n",
       "      <td>NBL1</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>MICOS10-NBL1</td>\n",
       "      <td>100532736.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>180 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       ensembl_gene_id external_gene_name external_gene_source   \n",
       "943    ENSG00000282566           PRAMEF25          HGNC Symbol  \\\n",
       "944    ENSG00000282566           PRAMEF25          HGNC Symbol   \n",
       "1038   ENSG00000114374              USP9Y          HGNC Symbol   \n",
       "1039   ENSG00000114374              USP9Y          HGNC Symbol   \n",
       "1196   ENSG00000228240            TTTY17A          HGNC Symbol   \n",
       "...                ...                ...                  ...   \n",
       "74592  ENSG00000229571           PRAMEF25          HGNC Symbol   \n",
       "75021  ENSG00000143702             CEP170          HGNC Symbol   \n",
       "75022  ENSG00000143702             CEP170          HGNC Symbol   \n",
       "75269  ENSG00000158747               NBL1          HGNC Symbol   \n",
       "75270  ENSG00000158747               NBL1          HGNC Symbol   \n",
       "\n",
       "      entrezgene_accession  entrezgene_id  \n",
       "943               PRAMEF25       441873.0  \n",
       "944               PRAMEF26       645359.0  \n",
       "1038                 USP9Y         8287.0  \n",
       "1039                TTTY15        64595.0  \n",
       "1196               TTTY17A       252949.0  \n",
       "...                    ...            ...  \n",
       "74592             PRAMEF26       645359.0  \n",
       "75021               CEP170         9859.0  \n",
       "75022             CEP170P1       645455.0  \n",
       "75269                 NBL1         4681.0  \n",
       "75270         MICOS10-NBL1    100532736.0  \n",
       "\n",
       "[180 rows x 5 columns]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "duplicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "211d6a3b-df7d-40a3-8e87-d1a32f66d779",
   "metadata": {},
   "outputs": [],
   "source": [
    "ensembl_duplicates = {}\n",
    "for index, row in duplicates.iterrows():\n",
    "    if row['ensembl_gene_id'] not in ensembl_duplicates.keys():\n",
    "        ensembl_duplicates[row['ensembl_gene_id']] = []\n",
    "    ensembl_duplicates[row['ensembl_gene_id']].append(row.entrezgene_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "0045d1b9-0fdf-40d7-a4c1-f22119999149",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "90"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(ensembl_duplicates.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "b79b0b60-86e0-4208-8345-af2d917d1f18",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ENSG00000282566': [441873.0, 645359.0],\n",
       " 'ENSG00000114374': [8287.0, 64595.0],\n",
       " 'ENSG00000228240': [252949.0, 474152.0],\n",
       " 'ENSG00000227439': [474151.0, 474152.0],\n",
       " 'ENSG00000222018': [101928147.0, 102723451.0],\n",
       " 'ENSG00000159216': [861.0, 100506403.0],\n",
       " 'ENSG00000205670': [54065.0, 102723553.0],\n",
       " 'ENSG00000285432': [245910.0, 503614.0],\n",
       " 'ENSG00000285832': [441317.0, 441327.0],\n",
       " 'ENSG00000284979': [245910.0, 503614.0],\n",
       " 'ENSG00000285015': [245908.0, 504180.0],\n",
       " 'ENSG00000168746': [140834.0, 101927242.0],\n",
       " 'ENSG00000288273': [255313.0, 728062.0],\n",
       " 'ENSG00000263247': [5554.0, 11272.0],\n",
       " 'ENSG00000282673': [5542.0, 653247.0],\n",
       " 'ENSG00000282269': [5554.0, 11272.0],\n",
       " 'ENSG00000275517': [5542.0, 653247.0],\n",
       " 'ENSG00000278632': [414059.0, 101060321.0],\n",
       " 'ENSG00000274166': [643707.0, 647042.0],\n",
       " 'ENSG00000282773': [727909.0, 728047.0],\n",
       " 'ENSG00000230417': [414243.0, 100132987.0],\n",
       " 'ENSG00000226725': [101928345.0, 101928380.0],\n",
       " 'ENSG00000160223': [23308.0, 102723996.0],\n",
       " 'ENSG00000275153': [101060351.0, 102723859.0],\n",
       " 'ENSG00000276349': [729873.0, 101060376.0],\n",
       " 'ENSG00000274939': [414059.0, 101060321.0],\n",
       " 'ENSG00000213809': [22914.0, 100528032.0],\n",
       " 'ENSG00000276299': [441873.0, 645359.0],\n",
       " 'ENSG00000281880': [440034.0, 103157000.0],\n",
       " 'ENSG00000063587': [10838.0, 105373378.0],\n",
       " 'ENSG00000276725': [9859.0, 645455.0],\n",
       " 'ENSG00000275003': [1652.0, 100037417.0],\n",
       " 'ENSG00000283348': [441873.0, 645359.0],\n",
       " 'ENSG00000253434': [105375706.0, 105375707.0],\n",
       " 'ENSG00000273772': [6606.0, 6607.0],\n",
       " 'ENSG00000276910': [2966.0, 728340.0],\n",
       " 'ENSG00000241863': [79897.0, 202658.0],\n",
       " 'ENSG00000275045': [2966.0, 728340.0],\n",
       " 'ENSG00000239865': [79897.0, 202658.0],\n",
       " 'ENSG00000277773': [6606.0, 6607.0],\n",
       " 'ENSG00000243009': [79897.0, 202658.0],\n",
       " 'ENSG00000254911': [619383.0, 100158262.0],\n",
       " 'ENSG00000241779': [79897.0, 202658.0],\n",
       " 'ENSG00000274729': [245908.0, 504180.0],\n",
       " 'ENSG00000281040': [441317.0, 441327.0],\n",
       " 'ENSG00000277530': [245910.0, 503614.0],\n",
       " 'ENSG00000292187': [4101.0, 266740.0],\n",
       " 'ENSG00000239927': [79897.0, 202658.0],\n",
       " 'ENSG00000242726': [79897.0, 202658.0],\n",
       " 'ENSG00000267452': [440446.0, 105371828.0],\n",
       " 'ENSG00000186854': [129293.0, 105374836.0],\n",
       " 'ENSG00000226023': [255313.0, 728062.0],\n",
       " 'ENSG00000099977': [1652.0, 100037417.0],\n",
       " 'ENSG00000257365': [2342.0, 100529261.0],\n",
       " 'ENSG00000189064': [729408.0, 729447.0],\n",
       " 'ENSG00000183305': [4101.0, 266740.0],\n",
       " 'ENSG00000004866': [7982.0, 93655.0],\n",
       " 'ENSG00000278662': [643707.0, 647042.0],\n",
       " 'ENSG00000241370': [79897.0, 202658.0],\n",
       " 'ENSG00000101883': [158800.0, 104797536.0],\n",
       " 'ENSG00000178115': [727909.0, 728047.0],\n",
       " 'ENSG00000239149': [677882.0, 677885.0],\n",
       " 'ENSG00000225194': [100188953.0, 105376159.0],\n",
       " 'ENSG00000186599': [245908.0, 504180.0],\n",
       " 'ENSG00000198129': [245910.0, 503614.0],\n",
       " 'ENSG00000186572': [245910.0, 503614.0],\n",
       " 'ENSG00000285975': [441317.0, 441327.0],\n",
       " 'ENSG00000258537': [145438.0, 100874185.0],\n",
       " 'ENSG00000229140': [137196.0, 728724.0],\n",
       " 'ENSG00000223403': [378881.0, 100507257.0],\n",
       " 'ENSG00000111850': [57150.0, 63914.0],\n",
       " 'ENSG00000214900': [196913.0, 283551.0],\n",
       " 'ENSG00000145736': [2966.0, 728340.0],\n",
       " 'ENSG00000274512': [729873.0, 101060376.0],\n",
       " 'ENSG00000143384': [4170.0, 574406.0],\n",
       " 'ENSG00000224897': [401398.0, 101928283.0],\n",
       " 'ENSG00000185164': [283820.0, 102723728.0],\n",
       " 'ENSG00000205571': [6606.0, 6607.0],\n",
       " 'ENSG00000170667': [10156.0, 100271927.0],\n",
       " 'ENSG00000273513': [101060351.0, 102723859.0],\n",
       " 'ENSG00000111215': [5554.0, 11272.0],\n",
       " 'ENSG00000274808': [414059.0, 101060321.0],\n",
       " 'ENSG00000251655': [5542.0, 653247.0],\n",
       " 'ENSG00000245552': [100129203.0, 101929295.0],\n",
       " 'ENSG00000196242': [81472.0, 148824.0],\n",
       " 'ENSG00000228262': [104355287.0, 104355288.0],\n",
       " 'ENSG00000290614': [150527.0, 646743.0],\n",
       " 'ENSG00000229571': [441873.0, 645359.0],\n",
       " 'ENSG00000143702': [9859.0, 645455.0],\n",
       " 'ENSG00000158747': [4681.0, 100532736.0]}"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensembl_duplicates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd2ff2ac-d672-49b2-a5d0-9f5e713ee3b2",
   "metadata": {},
   "source": [
    "### Keep the one in the human adata object if it is"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "b998301d-e345-45e6-a197-dbf576ffe73c",
   "metadata": {},
   "outputs": [],
   "source": [
    "ids = ensembl_duplicates['ENSG00000282566']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "0ff6bee3-8ba4-4b7c-9498-88b1cf5713fd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "645359.0"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensembl_duplicates['ENSG00000282566'][1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "8226eec2-d5c1-4476-bd40-c8204535e1a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_removed = []\n",
    "for gene in ensembl_duplicates.keys():\n",
    "    ids = ensembl_duplicates[gene]\n",
    "    if ids[1] in human.var.entrez_id:\n",
    "        genes_removed.append(ensembl_duplicates[gene][2])\n",
    "    else :\n",
    "        genes_removed.append(ensembl_duplicates[gene][1])                  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "e5ba0254-8fad-40cc-8561-6639a1aea023",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "90"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(genes_removed) #Same length as the duplicates in human_ncbi_id df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "fbde3a2a-b660-484d-9835-3293902c74d4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[645359.0,\n",
       " 64595.0,\n",
       " 474152.0,\n",
       " 474152.0,\n",
       " 102723451.0,\n",
       " 100506403.0,\n",
       " 102723553.0,\n",
       " 503614.0,\n",
       " 441327.0,\n",
       " 503614.0,\n",
       " 504180.0,\n",
       " 101927242.0,\n",
       " 728062.0,\n",
       " 11272.0,\n",
       " 653247.0,\n",
       " 11272.0,\n",
       " 653247.0,\n",
       " 101060321.0,\n",
       " 647042.0,\n",
       " 728047.0,\n",
       " 100132987.0,\n",
       " 101928380.0,\n",
       " 102723996.0,\n",
       " 102723859.0,\n",
       " 101060376.0,\n",
       " 101060321.0,\n",
       " 100528032.0,\n",
       " 645359.0,\n",
       " 103157000.0,\n",
       " 105373378.0,\n",
       " 645455.0,\n",
       " 100037417.0,\n",
       " 645359.0,\n",
       " 105375707.0,\n",
       " 6607.0,\n",
       " 728340.0,\n",
       " 202658.0,\n",
       " 728340.0,\n",
       " 202658.0,\n",
       " 6607.0,\n",
       " 202658.0,\n",
       " 100158262.0,\n",
       " 202658.0,\n",
       " 504180.0,\n",
       " 441327.0,\n",
       " 503614.0,\n",
       " 266740.0,\n",
       " 202658.0,\n",
       " 202658.0,\n",
       " 105371828.0,\n",
       " 105374836.0,\n",
       " 728062.0,\n",
       " 100037417.0,\n",
       " 100529261.0,\n",
       " 729447.0,\n",
       " 266740.0,\n",
       " 93655.0,\n",
       " 647042.0,\n",
       " 202658.0,\n",
       " 104797536.0,\n",
       " 728047.0,\n",
       " 677885.0,\n",
       " 105376159.0,\n",
       " 504180.0,\n",
       " 503614.0,\n",
       " 503614.0,\n",
       " 441327.0,\n",
       " 100874185.0,\n",
       " 728724.0,\n",
       " 100507257.0,\n",
       " 63914.0,\n",
       " 283551.0,\n",
       " 728340.0,\n",
       " 101060376.0,\n",
       " 574406.0,\n",
       " 101928283.0,\n",
       " 102723728.0,\n",
       " 6607.0,\n",
       " 100271927.0,\n",
       " 102723859.0,\n",
       " 11272.0,\n",
       " 101060321.0,\n",
       " 653247.0,\n",
       " 101929295.0,\n",
       " 148824.0,\n",
       " 104355288.0,\n",
       " 646743.0,\n",
       " 645359.0,\n",
       " 645455.0,\n",
       " 100532736.0]"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_removed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "7ea88a3a-03a8-4d79-86b0-a533468b44b8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ncbi_ids[~ncbi_ids.entrezgene_id.isin(genes_removed)].duplicated().any()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "a3c5f9e6-a2dc-4e7b-8d95-1901450403a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "human_genes = ncbi_ids[~ncbi_ids.entrezgene_id.isin(genes_removed)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "ea2d30ee-7bad-47bc-8a7c-b95f47b52afe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "human_genes.ensembl_gene_id.duplicated().any()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "89338304-6308-497f-af94-4b1df82cddd1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "human_genes.entrezgene_id.duplicated().any()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "fbd132f2-576c-4016-9c2c-528cfe155dd4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>external_gene_name</th>\n",
       "      <th>external_gene_source</th>\n",
       "      <th>entrezgene_accession</th>\n",
       "      <th>entrezgene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSG00000198888</td>\n",
       "      <td>MT-ND1</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>ND1</td>\n",
       "      <td>4535.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENSG00000198763</td>\n",
       "      <td>MT-ND2</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>ND2</td>\n",
       "      <td>4536.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENSG00000198804</td>\n",
       "      <td>MT-CO1</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>COX1</td>\n",
       "      <td>4512.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ENSG00000228253</td>\n",
       "      <td>MT-ATP8</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>ATP8</td>\n",
       "      <td>4509.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>ENSG00000198938</td>\n",
       "      <td>MT-CO3</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>COX3</td>\n",
       "      <td>4514.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75704</th>\n",
       "      <td>ENSG00000142615</td>\n",
       "      <td>CELA2A</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>CELA2A</td>\n",
       "      <td>63036.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75705</th>\n",
       "      <td>ENSG00000215704</td>\n",
       "      <td>CELA2B</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>CELA2B</td>\n",
       "      <td>51032.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75706</th>\n",
       "      <td>ENSG00000132906</td>\n",
       "      <td>CASP9</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>CASP9</td>\n",
       "      <td>842.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75716</th>\n",
       "      <td>ENSG00000215695</td>\n",
       "      <td>RSC1A1</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>RSC1A1</td>\n",
       "      <td>6248.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75718</th>\n",
       "      <td>ENSG00000116786</td>\n",
       "      <td>PLEKHM2</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>PLEKHM2</td>\n",
       "      <td>23207.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17426 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       ensembl_gene_id external_gene_name external_gene_source   \n",
       "5      ENSG00000198888             MT-ND1          HGNC Symbol  \\\n",
       "9      ENSG00000198763             MT-ND2          HGNC Symbol   \n",
       "15     ENSG00000198804             MT-CO1          HGNC Symbol   \n",
       "20     ENSG00000228253            MT-ATP8          HGNC Symbol   \n",
       "22     ENSG00000198938             MT-CO3          HGNC Symbol   \n",
       "...                ...                ...                  ...   \n",
       "75704  ENSG00000142615             CELA2A          HGNC Symbol   \n",
       "75705  ENSG00000215704             CELA2B          HGNC Symbol   \n",
       "75706  ENSG00000132906              CASP9          HGNC Symbol   \n",
       "75716  ENSG00000215695             RSC1A1          HGNC Symbol   \n",
       "75718  ENSG00000116786            PLEKHM2          HGNC Symbol   \n",
       "\n",
       "      entrezgene_accession  entrezgene_id  \n",
       "5                      ND1         4535.0  \n",
       "9                      ND2         4536.0  \n",
       "15                    COX1         4512.0  \n",
       "20                    ATP8         4509.0  \n",
       "22                    COX3         4514.0  \n",
       "...                    ...            ...  \n",
       "75704               CELA2A        63036.0  \n",
       "75705               CELA2B        51032.0  \n",
       "75706                CASP9          842.0  \n",
       "75716               RSC1A1         6248.0  \n",
       "75718              PLEKHM2        23207.0  \n",
       "\n",
       "[17426 rows x 5 columns]"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "human_genes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "110bad61-6eb6-413e-a07f-f695637c6c6b",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Check entrez_id doublets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "93dd26de-04b9-41c8-913f-b7af78644b4f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>external_gene_name</th>\n",
       "      <th>external_gene_source</th>\n",
       "      <th>entrezgene_accession</th>\n",
       "      <th>entrezgene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1671</th>\n",
       "      <td>ENSG00000238074</td>\n",
       "      <td>TSPY9</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>TSPY9</td>\n",
       "      <td>728132.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2305</th>\n",
       "      <td>ENSG00000233803</td>\n",
       "      <td>TSPY4</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>TSPY4</td>\n",
       "      <td>728395.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2318</th>\n",
       "      <td>ENSG00000225560</td>\n",
       "      <td>FAM197Y8</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>FAM197Y8</td>\n",
       "      <td>105379268.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2319</th>\n",
       "      <td>ENSG00000228383</td>\n",
       "      <td>FAM197Y7</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>FAM197Y7</td>\n",
       "      <td>105379269.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2326</th>\n",
       "      <td>ENSG00000229549</td>\n",
       "      <td>TSPY8</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>TSPY8</td>\n",
       "      <td>728403.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74969</th>\n",
       "      <td>ENSG00000143554</td>\n",
       "      <td>SLC27A3</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>SLC27A3</td>\n",
       "      <td>11000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75021</th>\n",
       "      <td>ENSG00000143702</td>\n",
       "      <td>CEP170</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>CEP170</td>\n",
       "      <td>9859.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75194</th>\n",
       "      <td>ENSG00000080947</td>\n",
       "      <td>CROCCP3</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>CROCCP3</td>\n",
       "      <td>114819.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75376</th>\n",
       "      <td>ENSG00000160767</td>\n",
       "      <td>ENTREP3</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>ENTREP3</td>\n",
       "      <td>10712.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75700</th>\n",
       "      <td>ENSG00000160818</td>\n",
       "      <td>GPATCH4</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>GPATCH4</td>\n",
       "      <td>54865.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2117 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       ensembl_gene_id external_gene_name external_gene_source   \n",
       "1671   ENSG00000238074              TSPY9          HGNC Symbol  \\\n",
       "2305   ENSG00000233803              TSPY4          HGNC Symbol   \n",
       "2318   ENSG00000225560           FAM197Y8          HGNC Symbol   \n",
       "2319   ENSG00000228383           FAM197Y7          HGNC Symbol   \n",
       "2326   ENSG00000229549              TSPY8          HGNC Symbol   \n",
       "...                ...                ...                  ...   \n",
       "74969  ENSG00000143554            SLC27A3          HGNC Symbol   \n",
       "75021  ENSG00000143702             CEP170          HGNC Symbol   \n",
       "75194  ENSG00000080947            CROCCP3          HGNC Symbol   \n",
       "75376  ENSG00000160767            ENTREP3          HGNC Symbol   \n",
       "75700  ENSG00000160818            GPATCH4          HGNC Symbol   \n",
       "\n",
       "      entrezgene_accession  entrezgene_id  \n",
       "1671                 TSPY9       728132.0  \n",
       "2305                 TSPY4       728395.0  \n",
       "2318              FAM197Y8    105379268.0  \n",
       "2319              FAM197Y7    105379269.0  \n",
       "2326                 TSPY8       728403.0  \n",
       "...                    ...            ...  \n",
       "74969              SLC27A3        11000.0  \n",
       "75021               CEP170         9859.0  \n",
       "75194              CROCCP3       114819.0  \n",
       "75376              ENTREP3        10712.0  \n",
       "75700              GPATCH4        54865.0  \n",
       "\n",
       "[2117 rows x 5 columns]"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "human_genes[human_genes.entrezgene_id.duplicated()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "a69a210f-d8ed-4419-9dea-a29240c71a88",
   "metadata": {},
   "outputs": [],
   "source": [
    "entrez_dup = human_genes[human_genes.entrezgene_id.isin(human_genes[human_genes.entrezgene_id.duplicated()].entrezgene_id)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "0e48590d-7b89-49ac-a137-e05ffa7a587f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>external_gene_name</th>\n",
       "      <th>external_gene_source</th>\n",
       "      <th>entrezgene_accession</th>\n",
       "      <th>entrezgene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>934</th>\n",
       "      <td>ENSG00000282584</td>\n",
       "      <td>PRAMEF4</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>PRAMEF4</td>\n",
       "      <td>400735.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>935</th>\n",
       "      <td>ENSG00000282663</td>\n",
       "      <td>PRAMEF10</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>PRAMEF10</td>\n",
       "      <td>343071.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>936</th>\n",
       "      <td>ENSG00000279195</td>\n",
       "      <td>PRAMEF7</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>PRAMEF7</td>\n",
       "      <td>441871.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>939</th>\n",
       "      <td>ENSG00000282119</td>\n",
       "      <td>PRAMEF6</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>PRAMEF6</td>\n",
       "      <td>440561.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>941</th>\n",
       "      <td>ENSG00000282437</td>\n",
       "      <td>PRAMEF27</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>PRAMEF27</td>\n",
       "      <td>101929983.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74969</th>\n",
       "      <td>ENSG00000143554</td>\n",
       "      <td>SLC27A3</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>SLC27A3</td>\n",
       "      <td>11000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75021</th>\n",
       "      <td>ENSG00000143702</td>\n",
       "      <td>CEP170</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>CEP170</td>\n",
       "      <td>9859.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75194</th>\n",
       "      <td>ENSG00000080947</td>\n",
       "      <td>CROCCP3</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>CROCCP3</td>\n",
       "      <td>114819.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75376</th>\n",
       "      <td>ENSG00000160767</td>\n",
       "      <td>ENTREP3</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>ENTREP3</td>\n",
       "      <td>10712.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75700</th>\n",
       "      <td>ENSG00000160818</td>\n",
       "      <td>GPATCH4</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>GPATCH4</td>\n",
       "      <td>54865.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3194 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       ensembl_gene_id external_gene_name external_gene_source   \n",
       "934    ENSG00000282584            PRAMEF4          HGNC Symbol  \\\n",
       "935    ENSG00000282663           PRAMEF10          HGNC Symbol   \n",
       "936    ENSG00000279195            PRAMEF7          HGNC Symbol   \n",
       "939    ENSG00000282119            PRAMEF6          HGNC Symbol   \n",
       "941    ENSG00000282437           PRAMEF27          HGNC Symbol   \n",
       "...                ...                ...                  ...   \n",
       "74969  ENSG00000143554            SLC27A3          HGNC Symbol   \n",
       "75021  ENSG00000143702             CEP170          HGNC Symbol   \n",
       "75194  ENSG00000080947            CROCCP3          HGNC Symbol   \n",
       "75376  ENSG00000160767            ENTREP3          HGNC Symbol   \n",
       "75700  ENSG00000160818            GPATCH4          HGNC Symbol   \n",
       "\n",
       "      entrezgene_accession  entrezgene_id  \n",
       "934                PRAMEF4       400735.0  \n",
       "935               PRAMEF10       343071.0  \n",
       "936                PRAMEF7       441871.0  \n",
       "939                PRAMEF6       440561.0  \n",
       "941               PRAMEF27    101929983.0  \n",
       "...                    ...            ...  \n",
       "74969              SLC27A3        11000.0  \n",
       "75021               CEP170         9859.0  \n",
       "75194              CROCCP3       114819.0  \n",
       "75376              ENTREP3        10712.0  \n",
       "75700              GPATCH4        54865.0  \n",
       "\n",
       "[3194 rows x 5 columns]"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "entrez_dup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "ca1a2d9e-9312-4dbd-b1b5-89a193b2902b",
   "metadata": {},
   "outputs": [],
   "source": [
    "entrezid = {}\n",
    "for index, row in entrez_dup.iterrows():\n",
    "    if row['entrezgene_id'] not in entrezid.keys():\n",
    "        entrezid[row['entrezgene_id']] = []\n",
    "    if row.ensembl_gene_id not in entrezid[row['entrezgene_id']]:\n",
    "        entrezid[row['entrezgene_id']].append(row.ensembl_gene_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "4c23ae8a-9fe8-4077-8590-163428b8d832",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1077"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(entrezid.keys())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd63b0c9-0ed6-4101-838c-5ecf351009cb",
   "metadata": {},
   "source": [
    "### Keep only the first"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "ba962eca-0938-40c9-bce7-bb6485b48795",
   "metadata": {},
   "outputs": [],
   "source": [
    "entrezgene_removed = []\n",
    "for entrez in entrezid.keys():\n",
    "    for gene in range(1,len(entrezid[entrez])):\n",
    "        entrezgene_removed.append(entrezid[entrez][gene])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "63c684a6-9560-4349-82bf-323a84f41583",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2117"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(entrezgene_removed) #Same length as the duplicates in human_ncbi_id df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "a9213e3e-cd27-4335-80cf-e9ea7d3f29b4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "human_genes[~human_genes.ensembl_gene_id.isin(entrezgene_removed)].entrezgene_id.duplicated().any()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "id": "3668fa9a-d9e2-484b-8353-9bf031263115",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_df = human_genes[~human_genes.ensembl_gene_id.isin(entrezgene_removed)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "id": "f2585025-2818-4bc8-8c79-655a92c1867c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 157,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_df.ensembl_gene_id.duplicated().any()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "id": "1cb22dbe-870f-4f36-8c95-d758b3e6eb5f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 158,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_df.entrezgene_id.duplicated().any()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "id": "a4aa99be-9566-4f2d-bb13-516602e961e6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>external_gene_name</th>\n",
       "      <th>external_gene_source</th>\n",
       "      <th>entrezgene_accession</th>\n",
       "      <th>entrezgene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSG00000198888</td>\n",
       "      <td>MT-ND1</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>ND1</td>\n",
       "      <td>4535.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENSG00000198763</td>\n",
       "      <td>MT-ND2</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>ND2</td>\n",
       "      <td>4536.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENSG00000198804</td>\n",
       "      <td>MT-CO1</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>COX1</td>\n",
       "      <td>4512.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ENSG00000228253</td>\n",
       "      <td>MT-ATP8</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>ATP8</td>\n",
       "      <td>4509.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>ENSG00000198938</td>\n",
       "      <td>MT-CO3</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>COX3</td>\n",
       "      <td>4514.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75704</th>\n",
       "      <td>ENSG00000142615</td>\n",
       "      <td>CELA2A</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>CELA2A</td>\n",
       "      <td>63036.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75705</th>\n",
       "      <td>ENSG00000215704</td>\n",
       "      <td>CELA2B</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>CELA2B</td>\n",
       "      <td>51032.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75706</th>\n",
       "      <td>ENSG00000132906</td>\n",
       "      <td>CASP9</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>CASP9</td>\n",
       "      <td>842.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75716</th>\n",
       "      <td>ENSG00000215695</td>\n",
       "      <td>RSC1A1</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>RSC1A1</td>\n",
       "      <td>6248.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75718</th>\n",
       "      <td>ENSG00000116786</td>\n",
       "      <td>PLEKHM2</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>PLEKHM2</td>\n",
       "      <td>23207.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15309 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       ensembl_gene_id external_gene_name external_gene_source   \n",
       "5      ENSG00000198888             MT-ND1          HGNC Symbol  \\\n",
       "9      ENSG00000198763             MT-ND2          HGNC Symbol   \n",
       "15     ENSG00000198804             MT-CO1          HGNC Symbol   \n",
       "20     ENSG00000228253            MT-ATP8          HGNC Symbol   \n",
       "22     ENSG00000198938             MT-CO3          HGNC Symbol   \n",
       "...                ...                ...                  ...   \n",
       "75704  ENSG00000142615             CELA2A          HGNC Symbol   \n",
       "75705  ENSG00000215704             CELA2B          HGNC Symbol   \n",
       "75706  ENSG00000132906              CASP9          HGNC Symbol   \n",
       "75716  ENSG00000215695             RSC1A1          HGNC Symbol   \n",
       "75718  ENSG00000116786            PLEKHM2          HGNC Symbol   \n",
       "\n",
       "      entrezgene_accession  entrezgene_id  \n",
       "5                      ND1         4535.0  \n",
       "9                      ND2         4536.0  \n",
       "15                    COX1         4512.0  \n",
       "20                    ATP8         4509.0  \n",
       "22                    COX3         4514.0  \n",
       "...                    ...            ...  \n",
       "75704               CELA2A        63036.0  \n",
       "75705               CELA2B        51032.0  \n",
       "75706                CASP9          842.0  \n",
       "75716               RSC1A1         6248.0  \n",
       "75718              PLEKHM2        23207.0  \n",
       "\n",
       "[15309 rows x 5 columns]"
      ]
     },
     "execution_count": 159,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c4eca7b-f492-4eeb-804e-99ae5cbcca65",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Create new var df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "id": "58b31451-d13b-47b7-95a9-acb902c369ce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 76533 × 31551\n",
       "    obs: 'sample_name', 'exp_component_name', 'cluster_label', 'cluster_color', 'cluster_order', 'class_label', 'class_color', 'class_order', 'subclass_label', 'subclass_color', 'subclass_order', 'donor_sex_label', 'donor_sex_color', 'donor_sex_order', 'region_label', 'region_color', 'region_order', 'cortical_layer_label', 'cortical_layer_color', 'cortical_layer_order', 'cell_type_accession_label', 'cell_type_accession_color', 'cell_type_accession_order', 'cell_type_alias_label', 'cell_type_alias_color', 'cell_type_alias_order', 'cell_type_alt_alias_label', 'cell_type_alt_alias_color', 'cell_type_alt_alias_order', 'cell_type_designation_label', 'cell_type_designation_color', 'cell_type_designation_order', 'external_donor_name_label', 'external_donor_name_color', 'external_donor_name_order', 'specimen_type', 'full_genotype_label', 'outlier_call', 'outlier_type', 'species', 'n_genes', 'cell_type'\n",
       "    var: 'gene', 'chromosome', 'entrez_id', 'gene_name', 'mouse_homologenes', 'n_cells'\n",
       "    layers: 'raw_counts'"
      ]
     },
     "execution_count": 160,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "human"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "id": "e05396bb-a95c-4b08-9a60-14d70505693b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs × n_vars = 76533 × 15309\n",
       "    obs: 'sample_name', 'exp_component_name', 'cluster_label', 'cluster_color', 'cluster_order', 'class_label', 'class_color', 'class_order', 'subclass_label', 'subclass_color', 'subclass_order', 'donor_sex_label', 'donor_sex_color', 'donor_sex_order', 'region_label', 'region_color', 'region_order', 'cortical_layer_label', 'cortical_layer_color', 'cortical_layer_order', 'cell_type_accession_label', 'cell_type_accession_color', 'cell_type_accession_order', 'cell_type_alias_label', 'cell_type_alias_color', 'cell_type_alias_order', 'cell_type_alt_alias_label', 'cell_type_alt_alias_color', 'cell_type_alt_alias_order', 'cell_type_designation_label', 'cell_type_designation_color', 'cell_type_designation_order', 'external_donor_name_label', 'external_donor_name_color', 'external_donor_name_order', 'specimen_type', 'full_genotype_label', 'outlier_call', 'outlier_type', 'species', 'n_genes', 'cell_type'\n",
       "    var: 'gene', 'chromosome', 'entrez_id', 'gene_name', 'mouse_homologenes', 'n_cells'\n",
       "    layers: 'raw_counts'"
      ]
     },
     "execution_count": 162,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "human[:,human.var.entrez_id.isin(final_df.entrezgene_id)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "id": "61a632af-6de3-4505-b3c3-0b952179bfc3",
   "metadata": {},
   "outputs": [],
   "source": [
    "human_reduce = human[:,human.var.entrez_id.isin(final_df.entrezgene_id)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "id": "39734654-ab1e-4c41-bd34-9099b13f8471",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/nfs/research/irene/anaelle/miniconda3/lib/python3.10/site-packages/anndata/_core/anndata.py:782: UserWarning: \n",
      "AnnData expects .var.index to contain strings, but got values like:\n",
      "    [144571, 127550, 51146, 729522, 13]\n",
      "\n",
      "    Inferred to be: integer\n",
      "\n",
      "  value_idx = self._prep_dim_index(value.index, attr)\n"
     ]
    }
   ],
   "source": [
    "human_reduce.var = human_reduce.var.set_index('entrez_id')\n",
    "human_reduce.var['entrez_id'] = human_reduce.var.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "id": "febd68e9-341e-4229-98a0-712e25480aee",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>mouse_homologenes</th>\n",
       "      <th>n_cells</th>\n",
       "      <th>entrez_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>entrez_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>144571</th>\n",
       "      <td>A2M-AS1</td>\n",
       "      <td>12</td>\n",
       "      <td>A2M antisense RNA 1 (head to head)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>197</td>\n",
       "      <td>144571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127550</th>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>1</td>\n",
       "      <td>alpha 1,3-galactosyltransferase 2</td>\n",
       "      <td>A3galt2</td>\n",
       "      <td>158</td>\n",
       "      <td>127550</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51146</th>\n",
       "      <td>A4GNT</td>\n",
       "      <td>3</td>\n",
       "      <td>alpha-1,4-N-acetylglucosaminyltransferase</td>\n",
       "      <td>A4gnt</td>\n",
       "      <td>17</td>\n",
       "      <td>51146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>729522</th>\n",
       "      <td>AACSP1</td>\n",
       "      <td>5</td>\n",
       "      <td>acetoacetyl-CoA synthetase pseudogene 1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10489</td>\n",
       "      <td>729522</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>AADAC</td>\n",
       "      <td>3</td>\n",
       "      <td>arylacetamide deacetylase</td>\n",
       "      <td>Aadac</td>\n",
       "      <td>545</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7789</th>\n",
       "      <td>ZXDA</td>\n",
       "      <td>X</td>\n",
       "      <td>zinc finger, X-linked, duplicated A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11839</td>\n",
       "      <td>7789</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158586</th>\n",
       "      <td>ZXDB</td>\n",
       "      <td>X</td>\n",
       "      <td>zinc finger, X-linked, duplicated B</td>\n",
       "      <td>Zxdb</td>\n",
       "      <td>64061</td>\n",
       "      <td>158586</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79699</th>\n",
       "      <td>ZYG11B</td>\n",
       "      <td>1</td>\n",
       "      <td>zyg-11 family member B, cell cycle regulator</td>\n",
       "      <td>Zyg11b</td>\n",
       "      <td>35581</td>\n",
       "      <td>79699</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7791</th>\n",
       "      <td>ZYX</td>\n",
       "      <td>7</td>\n",
       "      <td>zyxin</td>\n",
       "      <td>Zyx</td>\n",
       "      <td>1950</td>\n",
       "      <td>7791</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26009</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>1</td>\n",
       "      <td>zinc finger, ZZ-type containing 3</td>\n",
       "      <td>Zzz3</td>\n",
       "      <td>65196</td>\n",
       "      <td>26009</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15309 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              gene chromosome                                     gene_name   \n",
       "entrez_id                                                                     \n",
       "144571     A2M-AS1         12            A2M antisense RNA 1 (head to head)  \\\n",
       "127550     A3GALT2          1             alpha 1,3-galactosyltransferase 2   \n",
       "51146        A4GNT          3     alpha-1,4-N-acetylglucosaminyltransferase   \n",
       "729522      AACSP1          5       acetoacetyl-CoA synthetase pseudogene 1   \n",
       "13           AADAC          3                     arylacetamide deacetylase   \n",
       "...            ...        ...                                           ...   \n",
       "7789          ZXDA          X           zinc finger, X-linked, duplicated A   \n",
       "158586        ZXDB          X           zinc finger, X-linked, duplicated B   \n",
       "79699       ZYG11B          1  zyg-11 family member B, cell cycle regulator   \n",
       "7791           ZYX          7                                         zyxin   \n",
       "26009         ZZZ3          1             zinc finger, ZZ-type containing 3   \n",
       "\n",
       "          mouse_homologenes  n_cells  entrez_id  \n",
       "entrez_id                                        \n",
       "144571                  NaN      197     144571  \n",
       "127550              A3galt2      158     127550  \n",
       "51146                 A4gnt       17      51146  \n",
       "729522                  NaN    10489     729522  \n",
       "13                    Aadac      545         13  \n",
       "...                     ...      ...        ...  \n",
       "7789                    NaN    11839       7789  \n",
       "158586                 Zxdb    64061     158586  \n",
       "79699                Zyg11b    35581      79699  \n",
       "7791                    Zyx     1950       7791  \n",
       "26009                  Zzz3    65196      26009  \n",
       "\n",
       "[15309 rows x 6 columns]"
      ]
     },
     "execution_count": 165,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "human_reduce.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "id": "5523f91a-ceb9-42e3-bb66-518c8ea247c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_df = final_df.set_index('entrezgene_id')\n",
    "final_df['entrez_id'] = final_df.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "id": "df34c9e9-fa3b-4a4c-b220-be6ed1e2c03d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index([  4535.0,   4536.0,   4512.0,   4509.0,   4514.0,   4539.0,   4538.0,\n",
       "         4540.0,   4541.0, 400735.0,\n",
       "       ...\n",
       "         5362.0,   6262.0,   4595.0,   1441.0,   4610.0,  63036.0,  51032.0,\n",
       "          842.0,   6248.0,  23207.0],\n",
       "      dtype='float64', name='entrezgene_id', length=15309)"
      ]
     },
     "execution_count": 167,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_df.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "id": "e0269575-f13b-464e-8baf-a5fea22ca4b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "new_df = pd.concat([human_reduce.var, final_df], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "id": "9fa60f16-af31-4401-b581-8f0de7162823",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>mouse_homologenes</th>\n",
       "      <th>n_cells</th>\n",
       "      <th>entrez_id</th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>external_gene_name</th>\n",
       "      <th>external_gene_source</th>\n",
       "      <th>entrezgene_accession</th>\n",
       "      <th>entrez_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>144571.0</th>\n",
       "      <td>A2M-AS1</td>\n",
       "      <td>12</td>\n",
       "      <td>A2M antisense RNA 1 (head to head)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>197</td>\n",
       "      <td>144571</td>\n",
       "      <td>ENSG00000245105</td>\n",
       "      <td>A2M-AS1</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>A2M-AS1</td>\n",
       "      <td>144571.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127550.0</th>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>1</td>\n",
       "      <td>alpha 1,3-galactosyltransferase 2</td>\n",
       "      <td>A3galt2</td>\n",
       "      <td>158</td>\n",
       "      <td>127550</td>\n",
       "      <td>ENSG00000184389</td>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>127550.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51146.0</th>\n",
       "      <td>A4GNT</td>\n",
       "      <td>3</td>\n",
       "      <td>alpha-1,4-N-acetylglucosaminyltransferase</td>\n",
       "      <td>A4gnt</td>\n",
       "      <td>17</td>\n",
       "      <td>51146</td>\n",
       "      <td>ENSG00000118017</td>\n",
       "      <td>A4GNT</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>A4GNT</td>\n",
       "      <td>51146.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>729522.0</th>\n",
       "      <td>AACSP1</td>\n",
       "      <td>5</td>\n",
       "      <td>acetoacetyl-CoA synthetase pseudogene 1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10489</td>\n",
       "      <td>729522</td>\n",
       "      <td>ENSG00000291019</td>\n",
       "      <td>AACSP1</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>AACSP1</td>\n",
       "      <td>729522.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13.0</th>\n",
       "      <td>AADAC</td>\n",
       "      <td>3</td>\n",
       "      <td>arylacetamide deacetylase</td>\n",
       "      <td>Aadac</td>\n",
       "      <td>545</td>\n",
       "      <td>13</td>\n",
       "      <td>ENSG00000114771</td>\n",
       "      <td>AADAC</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>AADAC</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7789.0</th>\n",
       "      <td>ZXDA</td>\n",
       "      <td>X</td>\n",
       "      <td>zinc finger, X-linked, duplicated A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11839</td>\n",
       "      <td>7789</td>\n",
       "      <td>ENSG00000198205</td>\n",
       "      <td>ZXDA</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>ZXDA</td>\n",
       "      <td>7789.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158586.0</th>\n",
       "      <td>ZXDB</td>\n",
       "      <td>X</td>\n",
       "      <td>zinc finger, X-linked, duplicated B</td>\n",
       "      <td>Zxdb</td>\n",
       "      <td>64061</td>\n",
       "      <td>158586</td>\n",
       "      <td>ENSG00000198455</td>\n",
       "      <td>ZXDB</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>ZXDB</td>\n",
       "      <td>158586.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79699.0</th>\n",
       "      <td>ZYG11B</td>\n",
       "      <td>1</td>\n",
       "      <td>zyg-11 family member B, cell cycle regulator</td>\n",
       "      <td>Zyg11b</td>\n",
       "      <td>35581</td>\n",
       "      <td>79699</td>\n",
       "      <td>ENSG00000162378</td>\n",
       "      <td>ZYG11B</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>ZYG11B</td>\n",
       "      <td>79699.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7791.0</th>\n",
       "      <td>ZYX</td>\n",
       "      <td>7</td>\n",
       "      <td>zyxin</td>\n",
       "      <td>Zyx</td>\n",
       "      <td>1950</td>\n",
       "      <td>7791</td>\n",
       "      <td>ENSG00000285443</td>\n",
       "      <td>ZYX</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>ZYX</td>\n",
       "      <td>7791.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26009.0</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>1</td>\n",
       "      <td>zinc finger, ZZ-type containing 3</td>\n",
       "      <td>Zzz3</td>\n",
       "      <td>65196</td>\n",
       "      <td>26009</td>\n",
       "      <td>ENSG00000036549</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>26009.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15309 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             gene chromosome                                     gene_name   \n",
       "144571.0  A2M-AS1         12            A2M antisense RNA 1 (head to head)  \\\n",
       "127550.0  A3GALT2          1             alpha 1,3-galactosyltransferase 2   \n",
       "51146.0     A4GNT          3     alpha-1,4-N-acetylglucosaminyltransferase   \n",
       "729522.0   AACSP1          5       acetoacetyl-CoA synthetase pseudogene 1   \n",
       "13.0        AADAC          3                     arylacetamide deacetylase   \n",
       "...           ...        ...                                           ...   \n",
       "7789.0       ZXDA          X           zinc finger, X-linked, duplicated A   \n",
       "158586.0     ZXDB          X           zinc finger, X-linked, duplicated B   \n",
       "79699.0    ZYG11B          1  zyg-11 family member B, cell cycle regulator   \n",
       "7791.0        ZYX          7                                         zyxin   \n",
       "26009.0      ZZZ3          1             zinc finger, ZZ-type containing 3   \n",
       "\n",
       "         mouse_homologenes  n_cells  entrez_id  ensembl_gene_id   \n",
       "144571.0               NaN      197     144571  ENSG00000245105  \\\n",
       "127550.0           A3galt2      158     127550  ENSG00000184389   \n",
       "51146.0              A4gnt       17      51146  ENSG00000118017   \n",
       "729522.0               NaN    10489     729522  ENSG00000291019   \n",
       "13.0                 Aadac      545         13  ENSG00000114771   \n",
       "...                    ...      ...        ...              ...   \n",
       "7789.0                 NaN    11839       7789  ENSG00000198205   \n",
       "158586.0              Zxdb    64061     158586  ENSG00000198455   \n",
       "79699.0             Zyg11b    35581      79699  ENSG00000162378   \n",
       "7791.0                 Zyx     1950       7791  ENSG00000285443   \n",
       "26009.0               Zzz3    65196      26009  ENSG00000036549   \n",
       "\n",
       "         external_gene_name external_gene_source entrezgene_accession   \n",
       "144571.0            A2M-AS1          HGNC Symbol              A2M-AS1  \\\n",
       "127550.0            A3GALT2          HGNC Symbol              A3GALT2   \n",
       "51146.0               A4GNT          HGNC Symbol                A4GNT   \n",
       "729522.0             AACSP1          HGNC Symbol               AACSP1   \n",
       "13.0                  AADAC          HGNC Symbol                AADAC   \n",
       "...                     ...                  ...                  ...   \n",
       "7789.0                 ZXDA          HGNC Symbol                 ZXDA   \n",
       "158586.0               ZXDB          HGNC Symbol                 ZXDB   \n",
       "79699.0              ZYG11B          HGNC Symbol               ZYG11B   \n",
       "7791.0                  ZYX          HGNC Symbol                  ZYX   \n",
       "26009.0                ZZZ3          HGNC Symbol                 ZZZ3   \n",
       "\n",
       "          entrez_id  \n",
       "144571.0   144571.0  \n",
       "127550.0   127550.0  \n",
       "51146.0     51146.0  \n",
       "729522.0   729522.0  \n",
       "13.0           13.0  \n",
       "...             ...  \n",
       "7789.0       7789.0  \n",
       "158586.0   158586.0  \n",
       "79699.0     79699.0  \n",
       "7791.0       7791.0  \n",
       "26009.0     26009.0  \n",
       "\n",
       "[15309 rows x 11 columns]"
      ]
     },
     "execution_count": 179,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "id": "032f7d1c-d693-4469-826d-5303629735ff",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>mouse_homologenes</th>\n",
       "      <th>n_cells</th>\n",
       "      <th>entrez_id</th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>external_gene_name</th>\n",
       "      <th>external_gene_source</th>\n",
       "      <th>entrezgene_accession</th>\n",
       "      <th>entrez_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>entrez_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>144571</th>\n",
       "      <td>A2M-AS1</td>\n",
       "      <td>12</td>\n",
       "      <td>A2M antisense RNA 1 (head to head)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>197</td>\n",
       "      <td>144571</td>\n",
       "      <td>ENSG00000245105</td>\n",
       "      <td>A2M-AS1</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>A2M-AS1</td>\n",
       "      <td>144571.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127550</th>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>1</td>\n",
       "      <td>alpha 1,3-galactosyltransferase 2</td>\n",
       "      <td>A3galt2</td>\n",
       "      <td>158</td>\n",
       "      <td>127550</td>\n",
       "      <td>ENSG00000184389</td>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>127550.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51146</th>\n",
       "      <td>A4GNT</td>\n",
       "      <td>3</td>\n",
       "      <td>alpha-1,4-N-acetylglucosaminyltransferase</td>\n",
       "      <td>A4gnt</td>\n",
       "      <td>17</td>\n",
       "      <td>51146</td>\n",
       "      <td>ENSG00000118017</td>\n",
       "      <td>A4GNT</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>A4GNT</td>\n",
       "      <td>51146.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>729522</th>\n",
       "      <td>AACSP1</td>\n",
       "      <td>5</td>\n",
       "      <td>acetoacetyl-CoA synthetase pseudogene 1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10489</td>\n",
       "      <td>729522</td>\n",
       "      <td>ENSG00000291019</td>\n",
       "      <td>AACSP1</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>AACSP1</td>\n",
       "      <td>729522.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>AADAC</td>\n",
       "      <td>3</td>\n",
       "      <td>arylacetamide deacetylase</td>\n",
       "      <td>Aadac</td>\n",
       "      <td>545</td>\n",
       "      <td>13</td>\n",
       "      <td>ENSG00000114771</td>\n",
       "      <td>AADAC</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>AADAC</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7789</th>\n",
       "      <td>ZXDA</td>\n",
       "      <td>X</td>\n",
       "      <td>zinc finger, X-linked, duplicated A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11839</td>\n",
       "      <td>7789</td>\n",
       "      <td>ENSG00000198205</td>\n",
       "      <td>ZXDA</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>ZXDA</td>\n",
       "      <td>7789.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158586</th>\n",
       "      <td>ZXDB</td>\n",
       "      <td>X</td>\n",
       "      <td>zinc finger, X-linked, duplicated B</td>\n",
       "      <td>Zxdb</td>\n",
       "      <td>64061</td>\n",
       "      <td>158586</td>\n",
       "      <td>ENSG00000198455</td>\n",
       "      <td>ZXDB</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>ZXDB</td>\n",
       "      <td>158586.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79699</th>\n",
       "      <td>ZYG11B</td>\n",
       "      <td>1</td>\n",
       "      <td>zyg-11 family member B, cell cycle regulator</td>\n",
       "      <td>Zyg11b</td>\n",
       "      <td>35581</td>\n",
       "      <td>79699</td>\n",
       "      <td>ENSG00000162378</td>\n",
       "      <td>ZYG11B</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>ZYG11B</td>\n",
       "      <td>79699.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7791</th>\n",
       "      <td>ZYX</td>\n",
       "      <td>7</td>\n",
       "      <td>zyxin</td>\n",
       "      <td>Zyx</td>\n",
       "      <td>1950</td>\n",
       "      <td>7791</td>\n",
       "      <td>ENSG00000285443</td>\n",
       "      <td>ZYX</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>ZYX</td>\n",
       "      <td>7791.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26009</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>1</td>\n",
       "      <td>zinc finger, ZZ-type containing 3</td>\n",
       "      <td>Zzz3</td>\n",
       "      <td>65196</td>\n",
       "      <td>26009</td>\n",
       "      <td>ENSG00000036549</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>26009.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15309 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              gene chromosome                                     gene_name   \n",
       "entrez_id                                                                     \n",
       "144571     A2M-AS1         12            A2M antisense RNA 1 (head to head)  \\\n",
       "127550     A3GALT2          1             alpha 1,3-galactosyltransferase 2   \n",
       "51146        A4GNT          3     alpha-1,4-N-acetylglucosaminyltransferase   \n",
       "729522      AACSP1          5       acetoacetyl-CoA synthetase pseudogene 1   \n",
       "13           AADAC          3                     arylacetamide deacetylase   \n",
       "...            ...        ...                                           ...   \n",
       "7789          ZXDA          X           zinc finger, X-linked, duplicated A   \n",
       "158586        ZXDB          X           zinc finger, X-linked, duplicated B   \n",
       "79699       ZYG11B          1  zyg-11 family member B, cell cycle regulator   \n",
       "7791           ZYX          7                                         zyxin   \n",
       "26009         ZZZ3          1             zinc finger, ZZ-type containing 3   \n",
       "\n",
       "          mouse_homologenes  n_cells  entrez_id  ensembl_gene_id   \n",
       "entrez_id                                                          \n",
       "144571                  NaN      197     144571  ENSG00000245105  \\\n",
       "127550              A3galt2      158     127550  ENSG00000184389   \n",
       "51146                 A4gnt       17      51146  ENSG00000118017   \n",
       "729522                  NaN    10489     729522  ENSG00000291019   \n",
       "13                    Aadac      545         13  ENSG00000114771   \n",
       "...                     ...      ...        ...              ...   \n",
       "7789                    NaN    11839       7789  ENSG00000198205   \n",
       "158586                 Zxdb    64061     158586  ENSG00000198455   \n",
       "79699                Zyg11b    35581      79699  ENSG00000162378   \n",
       "7791                    Zyx     1950       7791  ENSG00000285443   \n",
       "26009                  Zzz3    65196      26009  ENSG00000036549   \n",
       "\n",
       "          external_gene_name external_gene_source entrezgene_accession   \n",
       "entrez_id                                                                \n",
       "144571               A2M-AS1          HGNC Symbol              A2M-AS1  \\\n",
       "127550               A3GALT2          HGNC Symbol              A3GALT2   \n",
       "51146                  A4GNT          HGNC Symbol                A4GNT   \n",
       "729522                AACSP1          HGNC Symbol               AACSP1   \n",
       "13                     AADAC          HGNC Symbol                AADAC   \n",
       "...                      ...                  ...                  ...   \n",
       "7789                    ZXDA          HGNC Symbol                 ZXDA   \n",
       "158586                  ZXDB          HGNC Symbol                 ZXDB   \n",
       "79699                 ZYG11B          HGNC Symbol               ZYG11B   \n",
       "7791                     ZYX          HGNC Symbol                  ZYX   \n",
       "26009                   ZZZ3          HGNC Symbol                 ZZZ3   \n",
       "\n",
       "           entrez_id  \n",
       "entrez_id             \n",
       "144571      144571.0  \n",
       "127550      127550.0  \n",
       "51146        51146.0  \n",
       "729522      729522.0  \n",
       "13              13.0  \n",
       "...              ...  \n",
       "7789          7789.0  \n",
       "158586      158586.0  \n",
       "79699        79699.0  \n",
       "7791          7791.0  \n",
       "26009        26009.0  \n",
       "\n",
       "[15309 rows x 11 columns]"
      ]
     },
     "execution_count": 180,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_df.reindex(human_reduce.var.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "id": "8def8f75-73de-41be-93cf-55f9032ea3bf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "15309"
      ]
     },
     "execution_count": 182,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum(human_reduce.var.index == new_df.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "id": "a28d5e03-60ae-4c69-837f-702d1edb731d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/nfs/research/irene/anaelle/miniconda3/lib/python3.10/site-packages/anndata/_core/anndata.py:782: UserWarning: \n",
      "AnnData expects .var.index to contain strings, but got values like:\n",
      "    [144571.0, 127550.0, 51146.0, 729522.0, 13.0]\n",
      "\n",
      "    Inferred to be: floating\n",
      "\n",
      "  value_idx = self._prep_dim_index(value.index, attr)\n"
     ]
    }
   ],
   "source": [
    "human_reduce.var = new_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "812a81a2-f732-4535-830a-f7dc360b1015",
   "metadata": {},
   "source": [
    "# Fly query"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f475e90c-e411-408e-94bd-213356e05e85",
   "metadata": {},
   "outputs": [],
   "source": [
    "fly_query = server.datasets['dmelanogaster_gene_ensembl']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56aa3002-e0d2-433f-b202-b45bddabfd13",
   "metadata": {},
   "outputs": [],
   "source": [
    "fly_query.show_attributes()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f6efc3af-fcf1-46a1-b39e-aaf70687ced4",
   "metadata": {},
   "outputs": [],
   "source": [
    "columns = ['ensembl_gene_id','external_gene_name', 'external_gene_source']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "4eb75165-4ad4-4ede-8d55-38ddbaa7d53d",
   "metadata": {},
   "outputs": [],
   "source": [
    "response = fly_query.search({'attributes' : columns,\n",
    "                           'mart_instance' : 'ensembl'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "fced1670-50c3-4d7f-b167-9216ff7283fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "mapping_fly = pd.read_csv(response.url, sep='\\t', header=None, names=columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "3c5cc9f1-7fce-422b-ac4c-d9423233cf0a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>external_gene_name</th>\n",
       "      <th>external_gene_source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FBgn0029843</td>\n",
       "      <td>Nep1</td>\n",
       "      <td>FlyBase gene name</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FBgn0266278</td>\n",
       "      <td>lncRNA:CR44963</td>\n",
       "      <td>FlyBase gene name</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FBgn0026238</td>\n",
       "      <td>gus</td>\n",
       "      <td>FlyBase gene name</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>FBgn0035235</td>\n",
       "      <td>CG7879</td>\n",
       "      <td>FlyBase gene name</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FBti0018960</td>\n",
       "      <td>flea{}771</td>\n",
       "      <td>Ensembl Metazoa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24273</th>\n",
       "      <td>FBgn0038358</td>\n",
       "      <td>Ttc26</td>\n",
       "      <td>FlyBase gene name</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24274</th>\n",
       "      <td>FBtr0445469_df_nrg</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24275</th>\n",
       "      <td>FBgn0030989</td>\n",
       "      <td>Inx5</td>\n",
       "      <td>FlyBase gene name</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24276</th>\n",
       "      <td>FBgn0259851</td>\n",
       "      <td>Su(Ste):CR42420</td>\n",
       "      <td>FlyBase gene name</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24277</th>\n",
       "      <td>FBgn0030673</td>\n",
       "      <td>CG15601</td>\n",
       "      <td>FlyBase gene name</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>24278 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          ensembl_gene_id external_gene_name external_gene_source\n",
       "0             FBgn0029843               Nep1    FlyBase gene name\n",
       "1             FBgn0266278     lncRNA:CR44963    FlyBase gene name\n",
       "2             FBgn0026238                gus    FlyBase gene name\n",
       "3             FBgn0035235             CG7879    FlyBase gene name\n",
       "4             FBti0018960          flea{}771      Ensembl Metazoa\n",
       "...                   ...                ...                  ...\n",
       "24273         FBgn0038358              Ttc26    FlyBase gene name\n",
       "24274  FBtr0445469_df_nrg                NaN                  NaN\n",
       "24275         FBgn0030989               Inx5    FlyBase gene name\n",
       "24276         FBgn0259851    Su(Ste):CR42420    FlyBase gene name\n",
       "24277         FBgn0030673            CG15601    FlyBase gene name\n",
       "\n",
       "[24278 rows x 3 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping_fly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "c22479ec-0778-46b3-9789-5908b1fbb4bd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>external_gene_name</th>\n",
       "      <th>external_gene_source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3500</th>\n",
       "      <td>FBgn0267508</td>\n",
       "      <td>28SrRNA-Psi:CR45848</td>\n",
       "      <td>FlyBase gene name</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     ensembl_gene_id   external_gene_name external_gene_source\n",
       "3500     FBgn0267508  28SrRNA-Psi:CR45848    FlyBase gene name"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping_fly[mapping_fly.ensembl_gene_id == 'FBgn0267508']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "331cbbb1-8a4f-4c45-923e-85b86c358c24",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
